// unitTest/TestRunner/TestInitializer/TestSerie/index.js
/*

*/

var TestPrototype = require('../TestPrototype');

var TestSerie = TestPrototype.create({
	Watcher: require('watcher'),
	Test: require('../Test'),
	type: 'testSerie',
	name: 'Anonymous testSerie',
	isWatching: true, // watch for test change and rerun the test when file is changed
	path: null,
	moduleTest: null,
	module: null,
	imports: null,
	failedCount: null,
	collectFails: false, // the serie end when a test fails
	tests: null,

	init: function(path, moduleTest){
		this.path = path;
		this.moduleTest = moduleTest;
		this.module = moduleTest.module;
		this.imports = moduleTest.imports;
		if( this.isWatching ) this.watch();
	},

	watch: function(){
		this.Watcher.watch(this.path, this.change.bind(this), this.watchFilter.bind(this));
	},

	change: function(){
		// when the test is in created state, no need to restart anything
		if( this.state != 'created' ){
			this.emit('change');
			this.restart();
		}
	},

	watchFilter: function(){
		return true;
	},

	loadTests: function(){
		// tests com from module.exports
		return Object.keys(this.exports);
	},

	createTest: function(name){
		return this.Test.new(name, this.exports[name], this);
	},

	setupModule: function(){
		this.exports = require(this.path);
	},

	teardownModule: function(){
		delete require.cache[this.path];
		this.exports = null;
	},

	setup: function(){
		this.setupModule();
		this.tests = this.loadTests();
		this.failedCount = 0;
		this.index = 0;
		this.current = null;
	},

	teardown: function(){
		// something to do?
	},

	clean: function(){
		this.teardownModule();
		if( this.current ){
			this.current.stop();
		}
	},

	next: function(){
		if( this.index >= this.tests.length ){
			return null;
		}
		else{
			this.current = this.createTest(this.tests[this.index]);
			// en faisant ça je peux récup les tests fait avant
			// cependant lorsque je restart je sais pas pourquoi
			// mais j'apelle pas getSerie
			this.tests[this.index] = this.current;
			this.current.handler = this;
			this.index++;
			return this.current;
		}
	},

	nextTest: function(){
		if( this.next() ){
			this.current.start();
		}
		// all test passed with success or no test to run
		else{
			this.pass();
		}
	},

	handleEvent: function(e){
		if( e.type == 'test-end' ){
			if( e.target.failed ){
				this.failedCount++;
			}

			if( this.failedCount && !this.collectFails ){
				this.fail();
				//this.error(e.target.lastError);
			}
			else{
				this.nextTest();
			}
		}

		TestPrototype.handleEvent.call(this, e);
	},

	run: function(){
		this.nextTest();
	}
});

module.exports = TestSerie;
