// RawStream/index.js
var ComputedStream = require('computedStream');
var proto = require('proto');

var RawStream = proto.extend(ComputedStream, {
	decoders: require('./decoders'),
	deconverters: require('./deconverters'),

	encoding: 'indentity', // source encoding
	charset: 'utf8', // source charset
	outputEncoding: 'identity',
	outputCharset: 'utf8',

	supportEncoding: function(encoding){
		return encoding === 'identity' || encoding in this.decoders;
	},

	supportCharset: function(charset){
		return charset === 'utf8' || charset in this.deconverters;
	},

	decode: function(encoding){
		if( encoding != 'identity' && encoding != this.encoding ){
			if( this.supportEncoding(encoding) ){
				this.chain(this.decoders[encoding].call(this));
			}
			else{
				throw new Error('cannot decode ' + encoding );
			}
		}
	},

	deconvert: function(charset){
		if( charset != 'utf8' && charset != this.charset ){
			if( this.supportCharset(charset) ){
				this.chain(this.deconverters[charset].call(this));
			}
			else{
				throw new Error('cannot deconvert  ' + charset);
			}
		}
	},

	_resolve: function(){
		this.decode(this.outputEncoding);	
		this.deconvert(this.outputCharset);
		return ComputedStream.prototype._resolve.call(this);
	}
});

module.exports = RawStream;
