// nodetest/TestFileCollector/index.js
/*

TestFileLoader.load(modulePath) -> returns an array of test file for this module or false

*/

var TestFileCollector = {
	path: require('path'),
	fileSystem: require('fs'),
	testFolderName: 'test',
	testFileName: 'test.js',

	getModuleFromTestFile: function(testFilePath){
		var dirname = this.path.dirname(testFilePath);

		if( dirname === this.testFolderName ){
			return this.path.dirname(dirname);
		}

		return dirname;
	},

	filterFolder: function(modulePath){
		var parts = modulePath.split(this.path.sep);
		return parts[parts.length - 1] !== this.testFolderName;
	},

	fromFile: function(path){
		var testFile, dirname = this.path.dirname(path);

		testFile = dirname + this.path.sep + this.testFileName;
		if( this.fileSystem.existsSync(testFile) ){
			return testFile;
		}

		testFile = dirname + this.path.sep + this.path.basename(path, '.js') + '-test.js';
		if( this.fileSystem.existsSync(testFile) ){
			return testFile;
		}

		return false;
	},

	filterFile: function(testFile){
		return testFile.slice(-3) === '.js';
	},

	fromFolder: function(path){
		var testFolder = this.path.dirname(path) + this.path.sep + this.testFolderName, files;
		
		try{
			// any files in it is a test
			files = require('fs.extra').readdirSyncRecursive(testFolder).filter(this.filterFile, this);
		}
		catch(e){
			files = false;
		}
		
		return files;
	},

	collect: function(modulePath, fn, bind){
		var file, error, result;

		try{
			file = this.fromFile(modulePath);
			result = file ? [file] : this.fromFolder(modulePath) || [];
		}
		catch(e){
			error = e;
		}

		fn.call(bind, error, result);
	}
};

module.exports = TestFileCollector;
