// PropertyAccessor/Lexer.js
/*

TODO: faudrait que 'name['name'] marche au même titre que 'name["name"]'

*/

var Lexer = require('lexer').extend({
	specials: '.[]()',
	quote: '"',
	cache: {},
	
	isQuote: function(char){
		return char == this.quote;
	},

	writeToken: function(token, char){
		// begining of a token
		if( this.isAlphaNum(char) ){
			token.index = this.index;
			token.value = char + this.readAlphaNum();
			if( char == '(' && this.nextChar() == ')' ){
				// token.args = eval('['+ this.buffer.slice(this.index, this.getSeparatorIndex(')'))  +']''); 
				// et faire pareil pour les expression entre []
				// tention: gros problème de sécurité si j'autorise ça coté serveur
				token.args = [];
			}

			return this.BREAK;
		}

		// begining of a token in quote like ["ok"]
		if( this.isQuote(char) ){
			token.index = this.index;
			token.value = this.readWhile(this.isQuote, false);

			if( token.index != this.index && !this.isQuote(char) ){
				throw this.createError('Unterminated quote at ' + token.index);
			}

			return this.BREAK;
		}
	}
});

Lexer.Token = Lexer.Token.extend({
	args: null
});

module.exports = Lexer;
