// Router/middleware/module/index.js
var Module = require('./Module');

var headers = {
	module: 'x-module',
	resolve: 'x-resolve',
	resolveParent: 'x-resolve-parent'
};

module.exports = {
	Module: Module,

	handle: function(request, response, handler){
		var Path = require('path'), module;

		if( request.hasHeader(headers.resolve) ){
			path = request.getHeader(headers.resolveParent);
			module = Module.new(path);
			
			module.resolve(request.getHeader(headers.resolve)).then(
				function(filename){
					if( module.isClientReadable() ){
						response.end(filename);
					}
					else{
						response.writeHeader(503); // unauthorized
						response.end();
					}
				},
				function(error){
					handler.next(error);
				}
			);
		}
		else if( request.hasHeader(headers.module) ){
			path = request.getHeader(headers.module);
			module = Module.new(path);

			module.resolve(path).then(
				function(filename){
					if( module.isClientReadable() ){
						response.sendFile(module.filename);
					}
					else{
						response.writeHeader(503); // unauthorized
						response.end();
					}
				},
				function(error){
					handler.next(error);
				}
			);
		}
		else{
			handler.next();
		}
	},

	use: function(router, options){
		if( options ){
			if( options.rootFolder ){
				Module.prototype.ModuleResolver.rootPath = options.rootFolder;
			}
			if( options.clientFolder ){
				Module.prototype.publicFolder = options.clientFolder;
			}
		}		
	}
};
