// LogStream/Log/index.js
/*

*/

var ansi = require('ansi');
var StylizedTemplate = require('../StringTemplate/StylizedTemplate');

var Log = StylizedTemplate.create({
	stackTrace: require('stackTrace'),
	inspect: require('util').inspect,
	fileSystem: require('fs'),
	ansi: ansi,

	styles: {},
	formats: Object.create(ansi.formats),

	depth: 2,
	showHidden: false,
	useStyle: false,
	autoReturn: true,

	level: null,
	message: null,
	data: null,

	init: function(level, message, data, styles){
		this.level = level;
		this.message = message;

		StylizedTemplate.init.call(this, message);

		this.data = data;
		if( styles ) this.registerStyles(styles);
		this.exec(data);
	},

	addFormat: function(styles){
		if( typeof styles == 'string' ) styles = [styles];
		this.formats[format] = styles;
	},

	writeStyle: function(string, format){
		return this.ansi.writeFormat(string, format);
	},

	// TODO: rewrite this function
	createDebugMessage: function(message){
		/*
		var call, wrap, lineFile, lineNumber, lineContent, levelPrefix;

		call = this.stackTrace.get()[1];
		lineFile = call.getFileName();
		lineNumber = call.getFileNumber();
		lineContent = this.stackTrace.readFileLine(lineFile, lineNumber);
		lineContent = lineContent.replace(/\n|\r/, '');
		message = message.replace(/\n|\r/g, function(match){ return match + '{debug}: '; });

		wrap = '\
			{debug}: \n\
			{debug}: {lineContent} ({lineFile: {lineNumber}})\n\
			{debug}: {message}\n\
			{debug}: \
		';

		return this._render(wrap, {
			debug: 'debug',
			message: message,
			lineContent: lineContent,
			lineFile: lineFile,
			lineNumber: lineNumber
		});
		*/
		return message;
	},

	toString: function(){
		var message = this.message, level = this.level, string;

		if( typeof message != 'string' ){
			message = this.inspect(message, {
				showHidden: this.showHidden,
				depth: this.depth,
				colors: false,
				customInspect: true
			});
		}

		string = level + ':' + StylizedTemplate.toString.call(this);

		if( this.autoReturn ) string+= '\n';

		return string;
	},

	toStylizedString: function(){
		var message = this.message, level = this.level, string;

		if( typeof message != 'string' ){
			message = this.inspect(message, {
				showHidden: this.showHidden,
				depth: this.depth,
				colors: true,
				customInspect: true
			});
		}

		string =  this.stylizeValue(level, level) + ':' + StylizedTemplate.toStylizedString.call(this);

		if( this.autoReturn ) string+= '\n';

		return string;
	},

	stringify: function(value){
		if( typeof value == 'string' ) return value;
		try{
			return JSON.stringify(value);
		}
		catch(e){
			return '[Circular]';
		}
	},

	/* create a short version of the data

	to resume calling
	logger.log('info', 'name: {name}', {name: 'damien', age:10});
	must ignore the age property to shorten the data passed to the streams
	logger.log('info', 'name: {name}', {name: 'damien', age:10});
	*/
	subsetValues: function(values){
		var i = 0, j = values.length, subset = {};

		for(;i<j;i++){
			subset[this.expressions[i]] = this.stringify(values[i]);
		}

		return subset;
	},

	subsetObjectOwnKeys: function(object){
		var keys = Object.keys(object), i = 0, j = keys.length, key, result;
		
		if( j ){
			result = {};
			for(;i<j;i++){
				key= keys[i];
				result[key] = object[key];
			}
		}

		return result;
	},

	toJSON: function(){
		return {
			level: this.level,
			message: this.message,
			data: this.data ? this.subsetValues(this.values) : null,
			expressionStyles: this.subsetObjectOwnKeys(this.expressionStyles),
			styleRules: this.subsetObjectOwnKeys(this.styleRules)
		};	
	}
});

Log.registerFormats({
	//'log': {},
	'info': 'green',
	'help': 'cyan',
	'warn': ['yellow', 'bold'],
	'debug': ['grey', 'bold'],
	'error': ['red', 'bold']
});

module.exports = Log;
