// LogStream/Log/index.js
var ConsoleTemplate = require('./ConsoleTemplate');
var proto = require('proto');

var Log = proto.extend(ConsoleTemplate, {
	level: null,
	suffix: '\n',
	mergePropertyNames: ConsoleTemplate.prototype.mergePropertyNames.concat('level'),

	constructor: function(string, level){
		//if( typeof level != 'string' ) throw new TypeError('Log must define a level');
		this.level = level;
		ConsoleTemplate.call(this, string);
	},

	toString: function(){
		var output;

		output = this.level && this.level != 'log' ? this.level + ':' : '';
		output+= ConsoleTemplate.prototype.toString.call(this);

		return output;
	},

	toStylizedString: function(){
		var output;

		output = this.level && this.level != 'log' ? this.stylizeValue(this.level, this.level) + ': ' : '';
		output+= ConsoleTemplate.prototype.toStylizedString.call(this);

		return output;
	},

	toJSON: function(){
		var obj = ConsoleTemplate.prototype.toJSON.call(this);
		obj.level = this.level;
		return obj;
	}
});

Log.prototype.registerStyles({
	//'log': {},
	'info': ['green'],
	'help': ['cyan'],
	'warn': ['yellow', 'bold'],
	'debug': ['grey', 'bold'],
	'error': ['red', 'bold']
});

module.exports = Log;

/* rewrite this function
		stackTrace: require('stackTrace'),
	fileSystem: require('fs'),
	createDebugMessage: function(message){
		var call, wrap, lineFile, lineNumber, lineContent, levelPrefix;

		call = this.stackTrace.get()[1];
		lineFile = call.getFileName();
		lineNumber = call.getFileNumber();
		lineContent = this.stackTrace.readFileLine(lineFile, lineNumber);
		lineContent = lineContent.replace(/\n|\r/, '');
		message = message.replace(/\n|\r/g, function(match){ return match + '{debug}: '; });

		wrap = '\
			{debug}: \n\
			{debug}: {lineContent} ({lineFile: {lineNumber}})\n\
			{debug}: {message}\n\
			{debug}: \
		';

		return this._render(wrap, {
			debug: 'debug',
			message: message,
			lineContent: lineContent,
			lineFile: lineFile,
			lineNumber: lineNumber
		});
		return message;
	}
*/

/*
logger.log({
	foo: 'bar',
	name: 1
});

va créer un inspectlog, voir comment on fait pis go
en gros c'est un log spécial qui n'a pas de level, bref à voir, ça me soule ce cas particulier de merde
*/

/*
var proto = require('proto');
var InspectLog = proto.create({
	inspect: require('util').inspect,
	ansi: require('ansi'),
	depth: 2,
	showHidden: false,

	constructor: function(object, styles){
		this.object = object;
	},

	stringify: function(string, stylized){
		var self = this;

		return this.inspect(string, {
			showHidden: this.showHidden,
			depth: this.depth,
			stylize: function(value, type){
				return stylized ? value : self.ansi.writeFormat(value, type);
			},
			customInspect: true
		});		
	},

	toString: function(){
		return this.stringify(this.object);
	},

	toStylizedString: function(){
		return this.stringify(this.object, true);
	},

	toJSON: function(){
		return {
			object: this.object
		};
	}
});

InspectLog.ansi.register({
	'number': 'yellow',
	'string': 'green',
	'date': 'magenta',
	'regexp': 'red',
	'null': 'bold',
	'undefined': 'grey',
	'special': 'cyan',
	'name': 'inherit'
});
*/
