// Map/index.js
/*
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

https://github.com/eriwen/es6-map-shim/blob/master/es6-map-shim.js
*/

var Iterator = require('Iterator');
var Map = function(iterable){
	this._keys = [];
	this._values = [];

	if( iterable ){
		Iterator.forOf(iterable, function(parts){
			var key = parts[0], value = parts[1];

			if( this._keys.indexOf(key) === -1 ){
				this._keys.push(key);
				this._values.push(value);
			}
		}, this);
	}
};

var proto = {
	constructor: Map,

	indexOf: function(key){

	},
	
	get size(){
		return this.keys.length;
	},

	forEach: function(fn, bind){

	},

	entries: function(){

	},

	keys: function(){

	},

	values: function(){

	},

	set: function(){

	},

	get: function(){

	},

	delete: function(){

	},

	clear: function(){

	},

	toString: function(){
		return '[object Map]';
	},

	'@@toStringTag': function(){
		return 'Map';
	},

	'@@iterator': function(){
		return this.entries();
	}
};

Map.prototype = proto;

module.exports = Map;

var topLevelScope = typeof window == 'undefined' ? global : window;
if( typeof topLevelScope['Map'] == 'undefined' ){
	topLevelScope['Map'] = Map;
}
