// Collector/index.js
/*
*/

var proto = require('proto');
var Iterator = require('./Iterator');

var Collector = proto.createFrom(Iterator, {
	iterable: null,
	options: {
		iterator: null,
		filter: null,
		filterBind: null,
		map: null,
		mapBind: null
	},
	result: {done: false, value: undefined},
	generateIndex: false,

	init: function(iterable, options, generateIndex){
		this.iterable = iterable;
		this.iterator = iterable['@@iterator']();

		if( options ){
			this.options = options;
		}
		this.generateIndex = generateIndex;
		this.index = 0;
	},

	next: function(){
		var result, iterator, next, value, filter, filterBind, map, mapBind;

		result = this.result;
		filter = this.options.filter;
		filterBind = this.options.filterBind;
		map = this.options.map;
		mapBind = this.options.mapBind;

		iterator = this.iterable['@@iterator']();
		next = iterator.next();
		while( next.done === false ){
			value = next.value;

			if( filter === null || typeof filterBind == 'undefined' ? filter(value) : filter.call(filterBind, value) ){
				if( this.generateIndex === true ){
					value =  typeof iterator.index == 'number' ? iterator.index : this.index;
				}

				if( map ){
					value = map.call(mapBind, value, this.index);
				}

				result.done = false;
				result.value = value;				

				return result;
			}

			next = iterator.next();
			this.index++;
		}

		result.done = true;
		result.value = undefined;
		return result;
	},

	'@@iterator': function(){
		return this.new(this.iterable, this.options);
	},

	collect: function(typeName){
		return this['@@iterator']();
	},

	collectIndexes: function(){
		return this.new(this.iterable, this.options, true);
	},

	count: function(){
		return Array.from(this).length;
	}
});

module.exports = Collector;
