// unitTest/ModuleTestSerie/index.js
/*


*/

var TestSerie = require('../TestSerie');
var ModuleTestSerie = TestSerie.create({
	path: null, // path of a folder/module to test

	Test: require('../ModuleTest'),
	Collector: require('./ModuleCollector'),

	init: function(path){
		this.path = path;
		this.modulePaths = this.Collector.collect(this.path);
		this.checkBadSignature();
		TestSerie.init.call(this, this.createTestList());
	},

	checkBadSignature: function(){
		var badSignaturePaths = this.Collector.badSignaturePaths;

		if( badSignaturePaths.length ){
			badSignaturePaths = badSignaturePaths.filter(this.Test.Collector.filterFolder, this.Test.Collector);
			if( badSignaturePaths.length ) this.emit('badSignature', badSignaturePaths);
		}
	},

	testComparer: function(a, b){
		return a.level - b.level;
	},

	createTestList: function(){
		var modulePaths = this.modulePaths, i = 0, j = modulePaths.length, modulePath, moduleTest, tests = [];

		for(;i<j;i++){
			modulePath = modulePaths[i];
			moduleTest = this.createTest(modulePath);
			if( moduleTest.tests.length ){
				tests.push(moduleTest);
			}
		}

		return tests.sort(this.testComparer);
	},

	observe: function(fn, bind){
		this.forEach(function(test){
			test.observe(fn, bind);
		});
	}
});

module.exports = ModuleTestSerie;
