// unitTest/TestRunner/TestPrototype/index.js
/*

maybe add a skip(reason) method a skip the test under some conditions (reason is optionnal)
and a method skipIf(condition, reason)

*/

var proto = require('proto');

var TestPrototype = proto.create({
	type: 'testPrototype',
	state: 'created', // created, stopped, started, ended
	startTime: null,
	endTime: null,
	timeout: null,
	failed: false,
	lastError: null,
	test: null,
	fn: null,
	bind: null,

	init: function(test){
		if( arguments.length === 0 ) throw new TypeError('test expected');
		if( typeof test != 'function' ) throw new TypeError('test must be a function');

		this.test = test;
	},

	clearTimer: function(){
		if( this.timer != null ){
			clearTimeout(this.timer);
			this.timer = null;
		}
	},

	// test results are ignored
	stop: function(){
		if( this.state == 'started' ){
			this.state = 'stopped';
			this.clearTimer();
			this.clean();
		}
	},

	_run: function(){
		return this.test.call(this, this);
	},

	run: function(fn, bind){
		this.fn = fn;
		this.bind = bind;
		return this._run();
	},

	// called before calling the test
	setup: function(){
		// noop
	},

	// called when the test has been called
	teardown: function(){
		// noop
	},

	// called to clean stuff when we want to rerun the tests while it was running
	clean: function(){
		// noop
	},

	start: function(){
		if( this.state == 'started' ){
			this.stop();
		}

		this.lastError = null;
		this.state = 'started';
		this.startTime = new Date().getTime();

		try{
			this.setup();
		}
		catch(e){
			return this.fail(e);
		}

		try{
			this.run();
		}
		catch(e){
			return this.fail(e);
		}

		if( typeof this.timeout == 'number' ){
			this.timer = setTimeout(this.onTimeOut.bind(this), this.timeout);
		}

		try{
			this.teardown();
		}
		catch(e){
			return this.fail(e);
		}
	},

	restart: function(){		
		if( this.state == 'started' ){
			this.stop();
		}
		else{
			this.clean();
		}

		this.start();
	},

	onEnd: function(){
		this.clean();

		if( this.state == 'ended' ){			
			if( this.failed ){
				this.fn.call(this.bind, this.lastError);
			}
			else{				
				this.fn.call(this.bind);
			}
		}
	},

	end: function(error){
		if( this.state == 'started' ){
			if( error ){
				this.failed = true;				
				this.lastError = error;
			}

			this.state = 'ended';
			this.endTime = new Date().getTime();
			this.clearTimer();
			//if( now ) this.onend();
			//else
			process.nextTick(this.onEnd.bind(this));
		}
	},

	fail: function(error){
		this.end(error);
	},

	pass: function(){
		this.end();
	},

	createTimeOutError: function(){
		var error = new Error('test takes too long');
		error.code = 'TEST_TIMEOUT';
		return error;
	},

	onTimeOut: function(){
		this.timer = null;
		this.fail(this.createTimeOutError());
	},

	get duration(){
		return this.endTime - this.startTime;
	}
});

module.exports = TestPrototype;
