// Object.instance/index.js
/*

Object.eachPair

provides: pair.eachPair, pair.eachPairIn,

*/

var pair = require('Object.pair');

if( 'getOwnPropertyNames' in Object ){

	pair.eachPair = function(object, fn, bind){
		var names = Object.getOwnPropertyNames(object), name, i, j, parentNames;

		i = 0;
		j = names.length;
		for(;i<j;i++){
			name = names[i];
			fn.call(bind, name, object[name], object);
		}

		while( object = Object.getPrototypeOf(object) ){
			if( object == Object.prototype ) break;
			parentNames = Object.getOwnPropertyNames(object);
			i = 0;
			j = parentNames.length;
			for(;i<j;i++){
				name = parentNames[i];
				if( names.indexOf(name) === -1 ){
					names.push(name);
					fn.call(bind, name, object[name], object);
				}
			}
		}
	};

	/*

	Object.getPropertyOwner = function(object, key){
		while( object ){
			if( Object.prototype.hasOwnProperty.call(object, key) ) return object;
			object = Object.getPrototypeOf(object);
		}
		return null;
	};

	Object.getPropertyDescriptor = function(object, key){
		object = Object.getPropertyOwner(object, key);
		return object ? Object.getOwnPropertyDescriptor(object, key) : null;
	};

	*/

}
else{

	pair.eachPair = function(object, fn, bind){
		for(var key in object) fn.call(bind, key, object[key], object);
		return object;
	};

}

pair.eachPairIn = function(array, fn, bind){
	return this.iteratePair(this.eachPair, array, fn, bind);
};

/*

Object.instance

provides: create, supplement, extend, new

*/

if( !Object.create ){
	Object.create = function(object){
		var F = function(){};
		F.prototype = object;
		return new F();
	};
}

module.exports = {
	create: function(){
		return Object.create(this);
	},

	new: function(){
		var instance = this.create(), init = instance.init;

		if( typeof init == "function" ) init.apply(instance, arguments);

		return instance;
	},

	supplement: function(){
		pair.eachPairIn(arguments, pair.appendPair, this);
		return this;
	},

	extend: function(){
		var object = this.create();

		object.supplement.apply(object, arguments);

		return object;
	},

	// allow to extend a function used as constructor
	extendConstructor: function(constructor){
		var proto = module.exports.create.call(constructor.prototype);

		proto.new = function(){
			var instance = this.create(), init = instance.init;

			constructor.apply(instance, arguments);

			if( typeof init == "function" ) init.apply(instance, arguments);

			return instance;
		};
		proto.constructor = proto.new;

		module.exports.supplement.apply(proto, Array.prototype.slice(arguments, 1));	

		return proto;
	}
};

pair.implement(Object, module.exports);
pair.implement(Function, 'extend', module.exports.extendConstructor);
