// Router/middleware/cors/index.js
/*
headerAllowed.push('content-disposition');
headerAllowed.push('Content-Transfer-Encoding');
headerAllowed.push('Content-Length');
headerAllowed.push('accept-encoding');
headerAllowed.push('authorization');
headerAllowed.push('user-agent');
headerAllowed.push('host');
headerAllowed.push('connection');

https://github.com/mcavage/node-restify/blob/master/lib/plugins/cors.js

*/

module.exports = {
	name: 'cors',

	response: {
		CORSAllowedMethods: ['GET', 'POST', 'PUT', 'DELETE', 'OPTIONS'],
		CORSAllowedHeaders: ['X-Requested-With', 'content-type', 'accept'],

		sendCORS: function(){
			this.statusCode = 204;
			this.setHeaders({
				'access-control-allow-origin': this.request.getHeader('origin') || '*',
				'access-control-allow-methods': this.CORSAllowedMethods.join(', '),
				'access-control-allow-headers': this.CORSAllowedHeaders.join(', '),
				'access-control-max-age': 10, // Seconds
				'content-length': 0
			});
		}
	},

	handle: function(request, response, handler){
		if( request.is('options') ){
			response.sendCORS();
		}
		else{
			handler.next();
		}
	}
};
