// Router/middleware/cors/index.js
/*
headerAllowed.push('content-disposition');
headerAllowed.push('Content-Transfer-Encoding');
headerAllowed.push('Content-Length');
headerAllowed.push('accept-encoding');
headerAllowed.push('authorization');
headerAllowed.push('user-agent');
headerAllowed.push('host');
headerAllowed.push('connection');

https://github.com/mcavage/node-restify/blob/master/lib/plugins/cors.js

*/

module.exports = {
	response: {
		methodsAllowed: ['GET', 'POST', 'PUT', 'DELETE', 'OPTIONS'],
		headersAllowed: ['X-Requested-With', 'content-type', 'accept'],

		sendCORS: function(){
			this.populate({
				statusCode: 204,
				headers: {
					'access-control-allow-origin': this.request.getHeader('origin') || '*',
					'access-control-allow-methods': this.methodsAllowed.join(', '),
					'access-control-allow-headers': this.headersAllowed.join(', '),
					'access-control-max-age': 10, // Seconds
					'content-length': 0
				}
			});
		}
	},

	handle: function(request, response, handler){
		if( request.is('options') ){
			response.sendCORS();
		}
		else{
			handler.next();
		}
	}
};
