// PropertyAccessor/Property/index.js
var Property = {
	previousPart: null,
	nextPart: null,
	args: null,
	hasModel: false,
	model: null,

	init: function(name){
		if( typeof name == 'string' ) name = name.trim();
		this.name = name;
	},

	setModel: function(model){
		this.unsetModel();
		this.model = model;
		this.hasModel = true;
		this.propagate();
		return this;
	},

	propagate: function(){
		if( this.nextPart ){
			if( this.hasModel ){
				this.nextPart.setModel(this.value);
			}
			else{
				this.nextPart.unsetModel();
			}
		}
	},

	unsetModel: function(){
		if( this.hasModel ){
			this.propagate();
			this.model = null;
			this.hasModel = false;
		}
		return this;
	},

	has: function(){
		return this.hasModel && this.name in this.model;	
	},

	hasOwn: function(){
		return this.hasModel && Object.prototype.hasOwnProperty.call(this.model, this.name);
	},

	set value(value){
		if( this.hasModel ){
			if( this.model != null ) this.model[this.name] = value;
			this.propagate();
		}
	},

	get value(){
		if( this.hasModel ){
			if( this.args ) return this.model[this.name].apply(this.model, this.args);
			return this.model[this.name];
		}
		return undefined;
	},

	apply: function(bind, args){
		if( this.hasModel ){
			var fn = this.model[this.name];
			return typeof fn == 'function' ? fn.apply(bind, args) : undefined;
		}
		return undefined;		
	},

	call: function(bind){
		return this.apply(bind, Array.slice(arguments, 1));
	}
};

module.exports = Property;
