// ansi/index.js
var ansi = {
	rules: {
		// font style
		'bold': [1, 22],
		'italic': [3, 23],
		'underline': [4, 24],
		'inverse': [7, 27],
		// grayscale colors
		'white': [37, 39],
		'grey': [90, 39],
		'black': [30, 39],
		//colors
		'blue': [34, 39],
		'cyan': [36, 39],
		'green': [32, 39],
		'magenta': [35, 39],
		'red': [31, 39],
		'yellow': [33, 39],
		// backgrougds
		'bg-black': [40, 49],
		'bg-red': [41, 49],
		'bg-green': [42, 49],
		'bg-yellow': [43, 49],
		'bg-blue': [44, 49],
		'bg-magenta': [45, 49],
		'bg-cyan': [46, 49],
		'bg-white': [47, 49]
	},
	styleRules: {},

	supported: function(){
		var argv = process.argv;

		if( argv.indexOf('--no-color') !== -1 || argv.indexOf('--color=false') !== -1 ){
			return false;
		}

		if( argv.indexOf('--color') !== -1 || argv.indexOf('--color=true') !== -1 || argv.indexOf('--color=always') !== -1 ){
			return true;
		}

		if( process.stdout && !process.stdout.isTTY ){
			return false;
		}

		if( process.platform === 'win32' ){
			return true;
		}

		if( 'COLORTERM' in process.env ){
			return true;
		}

		if( process.env.TERM === 'dumb' ){
			return false;
		}

		if( /^screen|^xterm|^vt100|color|ansi|cygwin|linux/i.test(process.env.TERM) ){
			return true;
		}

		return false;
	},

	defineStyle: function(name, rules){
		if( typeof rules == 'string' ) rules = [rules];
		this.styleRules[name] = rules;
		if( !(name in this) ){
			this[name] = function(str){
				return this.setStyle(str, name);
			};
		}
	},

	stylize: function(string, ruleName){
		var rule;

		if( ruleName in this.rules ){
			rule = this.rules[ruleName];
			string = '\033[' + rule[0] + 'm' + str + '\033[' + rule[1] + 'm';
		}

		return string;
	},

	stylizeAll: function(str, rules){

		for(var key in rules){
			str = this.stylize(str, key);
		}

		return str;
	},

	setStyle: function(str, name){
		return this.stylizeAll(str, this.styleRules[name]);
	},

	removeStyles: function(str){
		return String(str).replace(/\033\[\d+m/g, '');
	}
};

Object.keys(ansi.rules).forEach(function(rule){
	ansi.defineStyle(rule, rule);
});

module.exports = ansi;
