// ansi/index.js
var ansi = {
	tags: {
		//styles
		bold: [1, 22],
		italic: [3, 23],
		underline: [4, 24],
		inverse: [7, 27],
		//grayscale
		white: [37, 39],
		grey: [90, 39],
		black: [30, 39],
		//colors
		blue: [34, 39],
		cyan: [36, 39],
		green: [32, 39],
		magenta: [35, 39],
		red: [31, 39],
		yellow: [33, 39]
	},
	styles: {},

	defineStyle: function(name, properties){
		if( typeof properties == 'string' ) properties = {color: properties};
		this.styles[name] = properties;
		if( !(name in this) ){
			this[name] = function(str){
				return this.setStyle(str, name);
			};
		}
	},

	stylize: function(str, name){
		var tag = this.tags[name];
		if( tag ){
			str = '\033[' + tag[0] + 'm' + str + '\033[' + tag[1] + 'm';			
		}
		return str;
	},

	stylizeAll: function(str, properties){
		var key;

		for(key in properties){
			str = this.stylize(str, properties[key]);
		}

		return str;
	},

	setStyle: function(str, name){
		return this.stylizeAll(str, this.styles[name]);
	},

	removeStyles: function(str){
		return String(str).replace(/\u001b\[\d+m/g, '');
	}
};

Object.keys(ansi.tags).forEach(function(tagName){
	ansi.defineStyle(tagName, {color: tagName});
});

module.exports = ansi;
