// mdv/Template/TemplateParser/AttributeLinker/TokenListLinker/index.js
require('Node.bind');

/*

name: TokenListLinker
description: link a node attribute to a model multiple values (with prefix and suffix)
example: <span>Hello {name}, you are {age} years old!</span>

*/
var Linker = require('./linker');

var TokenListLinker = Linker.extend({
	ComputedBinding: require('../binding/computedBinding'),
	name: 'TokenListLinker',
	nodeAttribute: null,
	tokens: null,

	init: function(nodeAttribute, tokens){
		this.nodeAttribute = nodeAttribute;
		this.tokens = tokens;
	},

	replacer: function(token, values, i){
		return i in values ? values[i] : this.StringTemplate.empty(token);
	},

	replaceAll: function(values){
		return this.StringTemplate.replace(this.tokens, values, this.replacer, this);
	},

	replaceOne: function(values, computedBinding){
		return this.StringTemplate.replaceOne(this.tokens, computedBinding.id, computedBinding.lastChange.value);
	},

	link: function(node, model){
		var tokens = this.tokens, i = 1, j = tokens.length, computedBinding;

		computedBinding = this.ComputedBinding.new();
		for(;i<j;i+=2){
			// observe part, and pass true to avoid resolve call for each observer call
			computedBinding.observe(i, model, tokens[i], true);
		}
		computedBinding.value = this.replaceAll(computedBinding.values);
		computedBinding.combinator = this.replaceOne;
		computedBinding.combinatorBind = this;

		node.bind(this.nodeAttribute, computedBinding, 'value');
	},

	unlink: function(node, model){
		node.unbind(this.nodeAttribute);
	},

	namedScope: function(path, replace){
		var i = 1, tokens = this.tokens, j = tokens.length;
		for(;i<j;i+=2){
			tokens[i] = this.getNamedScopePath(tokens[i], path, replace);
		}
	}
});

module.exports = TokenListLinker;
