// Router/Response/index.js
/*

helper to manipulate the serverResponse

inspired from
https://github.com/mcavage/node-restify/blob/master/lib/response.js

node response source : 
https://github.com/joyent/node/blob/master/lib/_http_server.js

HOW TO SEND STATUS

response.statusCode = 200;

HOW TO SEND STATUS MESSAGE

response.statusMessage = 'ok'; 

HOW TO SEND BODY (prefer A or B)

A) readableStream.pipe(response);
B) response.end([String | Buffer], [encoding], [callback]);
C) readableStream.pipe(response, {end: false}); response.end(); 
D) response.write(String | Buffer, [encoding], [callback]); response.end();

------------- WITH CONTENT NEGOTIATION MIDDLEWARE -------------------

HOW TO SEND LENGTH

A) response.length = 10; // never touched
B) response.autoLength = true; // computed from the body

HOW TO SEND ENCODING

A) response.encoding = 'gzip'; // force gzip (can throw error)
B) response.negotiateEncoding = true; // use preferred encoding (can throw if no encoding is supported)

*/

var Emitter = require('Emitter');
var proto = require('proto');

var Response = proto.extend(Emitter, {
	ServerResponse: require('http').ServerResponse.prototype,
	serverResponse: null,
	request: null,
	writeCalls: null,
	waitCount: 0,

	constructor: function(response){
		Emitter.call(this);
		this.serverResponse = response;

		response.write = this.responseWrite.bind(this);
		response.end = this.responseEnd.bind(this);
		response.on('error', this.error.bind(this));

		/*
		node.js does this automatically
		this.on('beforeSendingHeaders', function(){
			if( this.hasHeader('content-length') === false ){
				// there is a body without content-length header, tell that to the browser
				this.setHeader('transfer-encoding', 'chunked');
				this.setHeader('accept-ranges', 'bytes');
			}
		});
		*/
	},

	getSourceObject: function(){
		var response = this.serverResponse;

		response.write = this.ServerResponse.write;
		response.end = this.ServerResponse.end;
		//response.removeListener('error', this.onerror);

		return response;
	},

	responseWrite: function(chunk, encoding){
		if( !this.headersSent ){
			// IRRELEVANT?? AJAX request need a 200 statusCode to read the response
			if( this.statusCode < 200 && this.statusCode >= 400 && this.request.isAjax() ){
				this.setHeader('x-status-code', this.statusCode); // error status are put in a header
				this.statusCode = 200;
			}

			this.emit('beforeSendingHeaders');
		}

		this.emit('write', chunk, encoding);
		if( this.writeCalls ){
			this.writeCalls.push(arguments);
			return false;
		}

		return this.ServerResponse.write.apply(this.serverResponse, arguments);
	},

	responseEnd: function(chunk, encoding){
		if( arguments.length !== 0 ){
			this.write(chunk, encoding);
		}

		if( this.writeCalls ){
			this.emit('end');
			return false;
		}

		return this.ServerResponse.end.call(this.serverResponse);
	},

	wait: function(fn, bind){
		if( !this.writeCalls ){
			this.writeCalls = [];
		}

		this.waitCount++;
		fn.call(bind, function(){
			// ce qu'on devait faire est terminé on peut appelé end
			this.waitCount--;
			if( this.waitCount === 0 ){
				this.writeCalls.forEach(function(args){
					this.ServerResponse.write.apply(this, args);
				}, this);
				this.ServerResponse.end.call(this);
			}
		}.bind(this));
	},

	get headers(){
		return this.serverResponse._headers;
	},

	set headers(object){
		for(var key in object){
			this.setHeader(key, object[key]);
		}
	},

	error: function(error){
		this.emit('error', error);
	},

	hasHeader: function(name){
		return typeof this.getHeader(name) != 'undefined';
	},

	send: function(status, chunk, encoding, callback){
		this.statusCode = status;
		return this.end(chunk, encoding, callback);
	}
});

// methods
['writeContinue', 'writeHead', 'setTimeout', 'setHeader',
'getHeader', 'removeHeader', /*'write',*/'addTrailers', 'end'].forEach(function(name){
	proto.define(Response, name, function(){
		return this.serverResponse[name].apply(this.serverResponse, arguments);
	});
});

// properties
['statusCode', 'headersSent', 'sendDate', 'statusMessage'].forEach(function(name){
	proto.defineProperty(Response, name, {
		get: function(){
			return this.serverResponse[name];
		},

		set: function(value){
			this.serverResponse[name] = value;
		}
	});
});

module.exports = Response;
