// FileChangeEmitter/index.js
/*


*/

var Emitter = require('Emitter');

var Notifier = Emitter.Notifier.create({
	fileSystem: require('fs'),
	watcher: null,
	lastChangeTime: null,
	changeInterval: 100,

	handleEvent: function(){
		var now = Number(new Date());

		if( now - this.lastChangeTime > this.changeInterval ){
			this.lastChangeTime = now;
			this.notify(this.name);
		}
	},

	open: function(){
		this.lastChangeTime = Number(new Date()) - this.changeInterval;
		this.watcher = this.fileSystem.watch(this.name, {persistent: false}, this.handleEvent.bind(this));
	},

	close: function(){
		this.watcher.close();
		this.watcher = null;
	}
});

var FileChangeEmitter = Emitter.create({
	resolve: require('path').resolve,
	fileSystem: require('fs'),
	Notifier: Notifier,

	encode: function(path){
		return this.resolve(path);
	},

	read: function(path, filter, depth){
		var paths = [];

		if( Array.isArray(path) ){
			path.forEach(function(sub){
				paths = paths.concat(this.read(sub, filter, depth));
			}, this);

			return paths;
		}

		if( !this.fileSystem.existsSync(path) ){
			console.warn(path, 'does not exists');
			return paths;
		}

		var stat = this.fileSystem.statSync(path);
		if( !filter || filter(path, stat) ){
			if( stat.isDirectory() ){
				if( typeof depth !== 'number' ) depth = -1;
				if( depth !== 0 ){
					// le chemin du fichier faut que je rajoute le chemin du dossier devant realPath ou path.resolve()?
					this.fileSystem.readdirSync(path).forEach(function(name){
						paths = paths.concat(this.read(path + '/' + name, filter, depth--));
					}, this);
				}
			}
			else{
				paths.push(path);
			}
		}

		return paths;
	}
});

module.exports = FileChangeEmitter;
