// mdv/parser/attributeLinker.js
var DirectLinker = require('./directLinker');
var TokenLinker = require('./tokenLinker');
var TokenListLinker = require('./tokenListLinker');
var MustacheTemplate = require('stringTemplate').extend({
	open: '{{',
	close: '}}',
	cache: {}, // need it's own cache

	replacer: function(values, token, i){
		return i in values ? values[i] : this.empty(token);
	}
});

var AttributeLinker = {
	StringTemplate: MustacheTemplate,

	new: function(attributeName, attributeValue){
		var stringTemplate = this.StringTemplate.new(attributeValue);
		var tokens = stringTemplate.parse();
		var length = tokens.length;

		// no mustache at all
		if( length < 3 ){
			return false;
		}
		
		// "Hello {{name}}, you are {{age}} years old": more than one mustache
		if( length > 3 ){
			return TokenListLinker.new(attributeName, stringTemplate);
		}

		// "{name}": contains one mustache with no prefix and no suffix
		if( tokens[0].length === 0 && tokens[2].length === 0 ){
			return DirectLinker.new(attributeName, tokens[1]);
		}

		// "Hello: {name}": contains one mustache with a prefix and a suffix
		return TokenLinker.new(attributeName, tokens[1], tokens[0], tokens[2]);
	}
};

module.exports = AttributeLinker;
