// mdv/parser/index.js
var Linker = require('./linker');

var Parser = {
	parsers: [],

	collect: function(node, path){
		var found = [], parsers = this.parsers, i = 0, j = parsers.length;
		var nodeType = node.nodeType, linker, terminal = false;

		if( typeof path != 'string' ) path = '';

		for(;i<j;i++){
			linker = parsers[i].call(this, node);
			if( Linker.isLinker(linker) ){
				linker.path = path;
				found.push(linker);
				if( linker.terminal ) terminal = true;
			}
		}

		// keep searching linkers
		if( terminal === false && (nodeType == 1 || nodeType == 11) ){
			found = found.concat(this.collectChildNodes(node, path));
		}

		return found;
	},

	collectNodeList: function(nodeList, path){
		var found = [], i = 0, j = nodeList.length;

		if( typeof path != 'string' ){
			path = '';
		}
		else if( path !== '' ){
			path+= '.';
		}

		for(;i<j;i++){
			found = found.concat(this.collect(nodeList[i], path + i));
		}

		return found;
	},

	collectChildNodes: function(node, path){
		return this.collectNodeList(node.childNodes, path);
	},

	parse: function(element, descendantOnly){
		var found;

		if( descendantOnly ){
			found = this.collectChildNodes(element);
		}
		else{
			found = this.collect(element);
		}

		return found;
	},

	register: function(directive){
		this.parsers.push(directive);
	}
};

var AttributeLinker = require('./attributeLinker');
Parser.register(function parseTextContent(node){
	if( node.nodeType != 3 ) return false;

	return AttributeLinker.new('textContent', node.textContent);
});

var LinkerListLinker = require('./linkerListLinker');
Parser.register(function parseAttributes(node){
	if( node.nodeType != 1 ) return false;

	var attributes = node.attributes, i = 0, j = attributes.length, attr, linker, linkerlist = [];

	for(;i<j;i++){
		attr = attributes[i];
		linker = AttributeLinker.new(attr.name, attr.value);
		if( linker ){
			linkerlist.push(linker);
		}
	}

	if( linkerlist.length === 0 ) return false;
	if( linkerlist.length === 1 ) return linkerlist[0];
	return LinkerListLinker.new(linkerlist);
});

var SubTemplateLinker = require('./subTemplateLinker');
var mdv = require('mdv');
Parser.register(function parseSubTemplate(node){
	if( mdv.isTemplate(node) ){
		return SubTemplateLinker.new(node);
	}
});

module.exports = Parser;
