// property/clone/index.js
/*

name: Object.cloning

*/

var clone = {
	enumerator: Object.keys,
	extensibility: false,

	cloneProperty: function(name, clone, object, seen){
		clone[name] = this.cloneProperties(object[name], seen);
	},

	cloneProperties: function(object, seen){
		var target = object;

		if( typeof object === 'object' && object !== null ){
			if( !seen ) seen = [object];
			else if( seen.indexOf(object) > -1 ) return object;
			else seen.push(object);

			if( typeof object.clone === 'function' ){
				target = object.clone();
			}
			else{
				var names = this.enumerator(object), i = names.length;		

				if( Object.prototype.toString.call(object) === '[object Array]' ){
					target = new Array(object.length);
				}
				else{
					target = Object.create(Object.getPrototypeOf(object));
				}

				while(i--){
					this.cloneMethod.call(this, names[i], target, object, seen);
				}

				// only if es5
				if( this.extensibility ){
					if( !Object.isExtensible(object) ) Object.preventExtensions(target);
					if( Object.isSealed(object) ) Object.seal(target);
					if( Object.isFrozen(object) ) Object.freeze(target);
				}
			}
		}

		return target;
	},

	clonePrimitive: function(){
		return this;
	},

	cloneNative: function(){
		return new this.constructor(this.valueOf());
	}
};

String.prototype.clone = clone.clonePrimitive;
Number.prototype.clone = clone.clonePrimitive;
Boolean.prototype.clone = clone.clonePrimitive;
RegExp.prototype.clone = clone.cloneNative;
Date.prototype.clone = clone.cloneNative;

var property = require('property');

[String, Number, Boolean, RegExp, Date].forEach(function(constructor){
	property.implement(constructor, 'clone', constructor.prototype.clone);
});

clone.cloneProperties = clone.cloneProperties.bind(clone);
clone.cloneMethod = clone.cloneProperty;
if( 'getOwnPropertyDescriptor' in Object ){
	Object.assign(clone, {
		extensibility: true,
		enumerator: Object.getOwnPropertyNames,
		
		clonePropertyDescriptor: function(name, clone, object, seen){
			var descriptor = Object.getOwnPropertyDescriptor(object, name);
			
			if( 'value' in descriptor ){
				descriptor.value = this.cloneProperties(descriptor.value, seen);
			}
			
			Object.defineProperty(clone, name, descriptor);
		}
	});

	clone.cloneMethod = clone.clonePropertyDescriptor;
}

module.exports = clone;
