// PropertyAccessorObserver/index.js
/* 

https://api.dartlang.org/apidocs/channels/stable/#observe/observe.PathObserver

property accessor: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Member_Operators

*/
var PropertyAccessor = require('PropertyAccessor');
var PropertyAccessorObserver = PropertyAccessor.create({
	partPrototype: require('./PropertyObserver'),
	closed: false,

	init: function(path, model, listener, bind, id, suppressNotify){
		PropertyAccessor.init.call(this, path);

		if( arguments.length > 1 ){
			this.setModel(model);
			if( arguments.length > 2 ){
				this.setListener(listener, bind || this, id, suppressNotify);
			}
		}
	},

	setListener: function(){
		this.lastPart.setListener.apply(this.lastPart, arguments);
		return this;
	},

	close: function(){
		// closing firstPart auto close nextParts
		this.firstPart.close.apply(this.firstPart, arguments);
		return this;
	},

	get lastChange(){
		return this.lastPart.lastChange;
	}
});

module.exports = PropertyAccessorObserver;
