// unitTest/AppTester/ModuleTestCollector/index.js
var proto = require('proto');

var ModuleTestCollector = proto.create({
	path: require('path'),
	fileSystem: require('fs'),
	testFolderName: 'test',
	testFileName: 'test.js',

	filter: function(testFile){
		return testFile.slice(-3) === '.js';
	},

	filterFolder: function(modulePath){
		var parts = modulePath.split(this.path.sep);

		return parts[parts.length - 1] !== this.testFolderName;
	},

	collect: function(modulePath){
		var tests = [], dirname, testFolderPath, testPath, isDir;

		if( this.filterFolder(modulePath) ){
			dirname = this.path.dirname(modulePath);
			testPath = dirname + this.path.sep + this.testFileName;

			if( this.fileSystem.existsSync(testPath) ){
				tests.push(testPath);
			}
			else{
				testFolderPath = dirname + this.path.sep + this.testFolderName;
				try{
					isDir = this.fileSystem.statSync(testFolderPath).isDirectory();
				}
				catch(e){
					isDir = false;
				}

				if( isDir ){
					// any files in it is a test
					tests = this.fileSystem.readdirSync(testFolderPath).filter(this.filter, this).map(function(name){
						return testFolderPath + this.path.sep + name;
					}, this);
				}
			}
		}

		return tests;
		
	}
});

module.exports = ModuleTestCollector;
