// Router/middleware/responseLength/index.js
var LengthGenerator = {
	length: 0,

	ondata: function(chunk, encoding){
		this.length+= this.response.computeLength(chunk, encoding);
	},

	onend: function(){
		this.response.length = this.length;
	}
};

module.exports = {
	name: 'responseLength',
	response: {
		autoLength: true, // computeLength from body
		generatingLength: false,

		get length(){
			return this.hasHeader('content-length') ? parseInt(this.getHeader('content-length'), 10) : 0;
		},

		set length(length){
			this.setHeader('content-length', length);
		},

		computeLength: function(chunk, encoding){
			return Buffer.isBuffer(chunk) ? chunk.length : Buffer.byteLength(chunk, encoding);
		},

		generateLengthHeader: function(){
			if( this.hasBody() && false === this.hasHeader('content-length') && this.autoLength ){
				this.generatingLength = true;
				this.prefetch(LengthGenerator);
			}
		}
	},

	handle: function(request, response, handler){
		response.on('beforeSendingHeaders', response.generateLengthHeader);
		handler.next();
	}
};
