// database/jsonFilePartManager/jsonFilePart.js
/*

FilePart in JSON format

item: the part as a JavaScript Object

*/

var FilePart = require('../FilePartManager/filePart');
var JSONFilePart = FilePart.extend({
	item: null,

	setBuffer: function(buffer, encoding){
		this.bufferError = null;

		if( Buffer.isBuffer(buffer) || typeof buffer == 'string' ){
			try{
				this.item = this.parse(typeof buffer == 'string' ? buffer : buffer.toString(encoding));
			}
			catch(e){
				this.bufferError = e;
			}
		}
		else if( typeof buffer == 'object' ){
			this.item = buffer;

			try{
				buffer = this.stringify(buffer);
			}
			catch(e){
				this.bufferError = e;
			}
		}

		return FilePart.setBuffer.apply(this, arguments);
	},

	empty: function(){
		this.item = JSONFilePart.item;
		return FilePart.empty.call(this);
	},

	parse: function(data){
		var item = null;

		if( data.length ){
			item = JSON.parse(data);

			/*
			item = JSON.parse(data, , function(key, value){
				if( key == 'byte' || key == 'index' || key == 'data' ){
					return this.error('les propriétés index, byte et data sont réservé à la BDD');
				}
			}.bind(this));
			*/
		}

		return item;
	},

	stringify: function(item){
		var data = '';

		/*
		data = JSON.stringify(item, function(key, value){
			if( key == 'line' ) return undefined;
			return value;
		});
		*/

		data = JSON.stringify(item);

		return data;
	}
});

module.exports = JSONFilePart;
