// nodetest/Test/index.js
/*

document.define('test', )

*/

var TestPrototype = require('../TestPrototype');
var proto = require('proto');

var Test = proto.extend(TestPrototype, {
	//clone: require('property/clone').cloneProperties,
	type: 'test',
	name: 'Anonymous test',
	timeout: 100,
	_module: null,
	imports: null,
	assertions: null,
	expectedAssertions: null,
	assertMethods: {
		'ok': function(a){
			return Boolean(a);
		},

		'equal': function(a, b){
			return a == b;
		},

		'strictEqual': function(a, b){
			return a === b;
		},

		'typeOf': function(a, b){
			return typeof a === b;
		},

		'protoOf': function(a, b){
			return Object.getPrototypeOf(a) === b;
		},

		'error': function(a, b){
			try{
				a();
			}
			catch(e){
				if( b ){
					if( typeof b == 'string' ){
						return e.type.toLowerCase() === b.toLowerCase();
					}
					else if( typeof b == 'function' ){
						return e instanceof b;
					}
				}				
				return true;
			}
			return false;
		}
	},

	constructor: function(name, test){
		TestPrototype.call(this, test);
		this.name = String(name);
	},

	get module(){
		return this._module || (this.parent ? this.parent.module : null);
	},

	set module(value){
		this._module = module;
	},

	createImports: function(){
		// can be commented because module cache is invalidated to module.exports is safe
		//return this.clone(this.module.exports);
		return this.module ? this.module.exports : null;
	},

	setup: function(){
		this.imports = this.createImports();
		this.assertions = [];
		//global.imports = this.imports;
	},

	teardown: function(){
		//global.imports = null;
	},

	run: function(){
		return this.test.call(this, this, this.imports);
	},

	expect: function(length){
		this.expectedAssertions = length;
		return this;
	},

	createAssertionError: function(type, args){
		var error = new Error('failed assertion');
		error.name = 'AssertionError';
		error.type = type;
		error.args = args;
		return error;
	},

	assert: function(type, args){
		// test is stopped or ended, this assertion is ignored
		if( this.state != 'started' ) return;

		var assertion = {
			type: type,
			args: args,
			ok: this.assertMethods[type].apply(this, args)
		};

		this.assertions.push(assertion);

		if( !assertion.ok ){
			// le test échoue dès qu'une assertion échoue
			this.fail(this.createAssertionError(type, args));
		}
	},

	createAssertionsLengthError: function(){
		return new Error('Expect' + this.expectedAssertions + 'and got ' + this.assertions.length);
	},

	done: function(){
		// test is stopped or ended, done() is ignored
		if( this.running === false ){
			
		}
		// only happens when expect(length) is called and we haven't check length stuff
		else if( typeof this.expectedAssertions == 'number' && this.assertions.length != this.expectedAssertions ){
			this.fail(this.createAssertionsLengthError());
		}
		else{
			this.pass();
		}
	},

	toString: function(){
		return '[object Test ' + this.name + ']';
	}
});

Object.keys(Test.prototype.assertMethods).forEach(function(key){
	Test.prototype[key] = function(){
		return this.assert(key, arguments);
	};
});

module.exports = Test;
