// Emitter/index.js
/*

name: Emitter

description: Emitter provide methods to emit and listen for events

require: Array.slice

*/

var proto = require('proto');
require('core/array');

var Emitter = proto.create({
	bind: null,
	eventRegExp: /\s+/,
	$listeners: {},
	size: 0,
	
	init: function(bind){
		if( arguments.length > 0 ) this.bind = bind;
		this.$listeners = proto.create.call(this.$listeners);
	},

	listeners: function(name){
		var listeners = this.$listeners, list;

		if( name in listeners ) list = listeners[name];
		else list = false;

		return list;
	},

	addListener: function(name, listener){
		if( typeof name != 'string' && typeof name != 'number' ){
			throw new TypeError('name expect string or number, ' + name + ' given');
		}
		if( typeof listener != 'function' && typeof listener != 'object' ){
			throw new TypeError('listener expect function or object, ' + listener + ' given');
		}

		var listeners = this.$listeners;

		if( name in listeners ){
			listeners[name].push(listener);
		}
		else{
			listeners[name] = [listener];
		}

		this.size++;

		if( name != 'addListener' ) this.applyListeners('addListener', arguments);

		return this;
	},

	removeListener: function(name, listener){
		var listeners = this.$listeners, list, retain, i, j, item;

		if( name == null ){
			for(name in listeners) this.removeListener(name, listener);
		}
		else if( listener == null ){
			if( name in listeners ){
				[].concat(listeners[name]).forEach(function(listener){
					this.removeListener(name, listener);
				}, this);
			}
		}
		else if( name in listeners ){
			list = listeners[name];
			retain = listeners[name] = [];
			i = 0;
			j = list.length;

			for(;i<j;i++){
				item = list[i];

				if( item === listener || item.__listener === listener ){
					this.size--;
				}
				else{
					retain.push(item);
				}
			}
			if( retain.length === 0 ){
				delete listeners[name];
			}

			if( name != 'removeListener' ) this.applyListeners('removeListener', arguments);
		}

		return this;
	},

	addVolatileListener: function(name){
		var self = this, once, args = arguments;

		// transform the second argument (supposed to be a function)
		// into a function that remove herself before calling original function
		once = function(){
			self.removeListener.apply(self, args);
			return once.__listener.apply(this, arguments);
		};
		once.__listener = args[1];
		args[1] = once;

		return this.addListener.apply(this, args);
	},

	applyListener: function(listener, name, args){
		if( typeof listener == 'object' ){
			return listener.handleEvent(name, args, this);
		}
		else{
			return listener.apply(this.bind || this, args);
		}
	},

	applyListeners: function(name, args){
		var listeners = this.listeners(name), i, j;

		if( name != 'applyListeners' && name != 'addListener' && name != 'removeListener' && name != '*' ){
			this.applyListeners('*', arguments);
		}

		if( listeners ){
			i = 0;
			j = listeners.length;
			for(;i<j;i++){
				this.applyListener(listeners[i], name, args);
			}
		}

		return this;
	},

	callListeners: function(name){
		return this.applyListeners(name, Array.slice(arguments, 1));
	},

	/*

	implement multiple event writing style:
	on({focus: function(){}, blur: function(){}});
	off('focus blur');
	emit('focus blur', true);
	*/
	eachEvent: function(method, bind, args){
		var name = args[0], key;

		if( args.length === 0 ){
			method.call(bind);
		}
		else if( typeof name == 'string' ){
			if( this.eventRegExp.test(name) ){
				args[0] = name.split(this.eventRegExp);
				return this.eachEvent(method, args);
			}
			else{
				method.apply(bind, args);
			}
		}
		else if( typeof name == 'number' ){
			method.apply(bind, args);
		}
		else if( name instanceof Array ){
			var i = 0, j = name.length;
			for(;i<j;i++){
				args[0] = name[i];
				method.apply(bind, args);
			}
		}
		else if( typeof name == 'object' ){
			args = Array.slice(args, 1);
			for(key in name){
				method.apply(bind, [key, name[key]].concat(args));
			}
		}

		return this;
	},

	on: function(){
		return this.eachEvent(this.addListener, this, arguments);
	},

	off: function(){
		return this.eachEvent(this.removeListener, this, arguments);
	},

	once: function(){
		return this.eachEvent(this.addVolatileListener, this, arguments);
	},

	emit: function(){
		return this.eachEvent(this.callListeners, this, arguments);
	}
});

module.exports = Emitter;
