// Router/middleware/cookieParser/index.js
var Cookie = require('cookie');

module.exports = {
	request: {
		cookieParams: {},

		parseCookie: function(cookies){
			try{
				return Cookie.parseAll(cookies);
			}
			catch(e){
				return e;
			}
		},
	},

	response: {
		createCookie: function(){
			return Cookie.new.apply(Cookie, arguments);
		},

		addCookie: function(cookie){
			if( false === Cookie.isPrototypeOf(cookie) ){
				throw new TypeError('cookie expected');
			}

			var cookies = this.headers[this.HEADERS.SET_COOKIE];

			if( typeof cookies == 'string' ){
				cookies = [cookies];
			}
			else if( !Array.isArray(cookies) ){
				cookies = [];
			}

			cookies.push(cookie.toString());
			this.headers[this.HEADERS.SET_COOKIE] = cookies;
			return cookie;
		}
	},

	handle: function(router){
		var request = router.request, cookies = request.getHeader(router.HEADERS.COOKIE);

		if( cookies ){
			request.cookieParams = request.parseCookie(cookies);
			if( request.cookieParams instanceof Error ) return router.error(request.cookieParams);
		}
		
		router.next();
	}
};
