// router/middleware/logger/index.js
/*

Response is missing the connection keep alive and transfer encoding chunked headers
because of this
https://github.com/joyent/node/blob/master/lib/_http_outgoing.js#L261

*/

var StringTemplate = require('StringTemplate');
var ConsoleTemplate = require('LogStream/Log/ConsoleTemplate');
var TableTemplate = require('TableTemplate');
var debug = require('debug');

module.exports = {
	name: 'logger',

	extend: {
		logRequest: true,
		logRequestBody: true,
		logRequestBodyMaxLength: 1000,
		logRequestTemplate: '{method} {url} HTTP/{httpVersion}\n{headers}{body}',

		getRequestLogScope: function(){
			var request = this.request, table = new TableTemplate();

			table.addRows(Object.keys(request.headers || {}).map(function(name){
				return [name, request.headers[name]];
			}, this));
			table.className = 'headless';
			table.defineClass('headerName', ['bold', 'grey']);
			table.column(0).setWidth(20).setClass('headerName');

			var headerString = table.toString();

			return {
				method: request.method.toUpperCase(),
				url: String(request.url),
				httpVersion: request.httpVersion,
				headers: headerString
			};
		},

		logResponse: true,
		logResponseBody: true,
		logResponseBodyMaxLength: 1000,
		logResponseTemplate: '{status} {statusMessage} HTTP/1.1\n{headers}{body}',

		getStatusStyle: function(){
			var status = this.statusCode;

			if( status >= 500 ){
				return 'red';
			}
			if( status >= 400 ){
				return 'yellow';
			}
			if( status >= 300 ){
				return 'cyan';
			}
			if( status >= 200 ){
				return 'green';
			}
			return 'inherit';
		},

		getResponseLogScope: function(response){
			var table = new TableTemplate();

			table.addRows(Object.keys(response.headers).map(function(name){
				return [name, response.headers[name]];
			}, this));

			table.className = 'human';
			table.defineClass('headerName', ['bold', 'grey']);
			table.column(0).setWidth(20).setClass('headerName');
			table.sortRows(0, 1);

			var headerString = table.toString();

			return {
				status: response.status,
				statusMessage: require('http').STATUS_CODES[response.status] || 'unknown',
				headers: headerString
			};
		}
	},

	use: function(router, logger){
		if( logger ) router.logger = logger;
	},

	handle: function(handler){
		var request = handler.request, logger = handler.router.logger;

		if( handler.logResponse ){
			handler.addTask(function(response){
				debug('creating response log');

				var scope = handler.getResponseLogScope(response);

				return Promise.resolve(handler.logResponseBody ? response.createBodyPromise() : '').then(function(body){
					debug('response body resolved');

					if( body.length > handler.logResponseBodyMaxLength ){
						body = body.slice(0, handler.logResponseBodyMaxLength) + '...';
					}
					scope.body = body.length ? '\n' + body : '';

					logger.log(handler.logResponseTemplate, scope, {
						status: handler.getStatusStyle(response.status),
						time: 'grey',
						body: 'grey'
					});
					return response;
				});
			});
		}

		if( handler.logRequest ){
			if( handler.logRequestBody ){
				var scope = handler.getRequestLogScope();

				return Promise.resolve(handler.logRequestBody ? request.createBodyPromise() : '').then(function(body){
					if( body.length > handler.logRequestBodyMaxLength ){
						body = body.slice(0, handler.logRequestBodyMaxLength) + '...';
					}
					scope.body = body.length ? '\n' + body : '';

					logger.log(handler.logRequestTemplate, scope, {
						method : 'magenta'
					});
				});
			}
		}
	}
};
