// NodeProcess/index.js
var Emitter = require('Emitter');
var FileObserver = require('FileObserver');
var proto = require('proto');
var prepareArgs = require('prepareProcessArgs');

var NodeProcess = proto.extend(Emitter, {
	Path: require('path'),
	childProcess: require('child_process'),
	args: null,
	state: 'created', // created, started, killed, restarting
	process: null,
	ctime: null,
	hasRestartFile: false,
	restartFiles: null,
	console: console,

	constructor: function(path, args){
		Emitter.call(this);
		this.path = this.Path.normalize(path);
		this.args = args;
	},

	onFileChange: function(path){
		this.emit('fileChange', path);
		this.console.warn('nodeprocess restart - {processPath} ({path} modified)', {path: path, processPath: this.path});
		this.restart();
	},

	setRestartFiles: function(filePaths){
		this.restartFiles = filePaths;

		var extraFS = require('fs.extra');
		var i = 0, j = this.restartFiles.length;
		this.hasRestartFile = true;
		for(;i<j;i++){
			extraFS.collectFileSync(this.restartFiles[i]).forEach(function(file){
				FileObserver.prototype.observe(file, this.onFileChange, this);
			}, this);
		}
	},

	isWindows: function(){
		return process.platform === 'win32';
	},

	start: function(){
		if( this.state == 'started' || this.state == 'restarting' ){
			throw new TypeError('cannot start, wrong process state' + this.state);
		}

		var preparedArgs = prepareArgs(this.args);

		this.process = this.childProcess.fork(this.path, preparedArgs, {
			cwd: this.Path.dirname(this.path)
		});
		this.ctime = Number(new Date());

		this.process.on('exit', this.onexit.bind(this));
		this.process.on('message', this.onmessage.bind(this));

		this.state = 'started';
		this.emit('start');
	},

	restart: function(){
		if( this.state == 'started' ){
			this.state = 'restarting';
			if( this.isWindows() ){
				this.kill();
			}
			else{
				this.kill('SIGUSR2');
			}
		}
		else if( this.state != 'restarting' ){
			this.emit('restart');
			this.start();
		}
	},

	kill: function(signal){
		this.emit('kill', signal);
		this.process.kill(signal);
	},

	onexit: function(code, signal){
		if( this.state == 'restarting' ) signal = 'SIGUSR2';
		// this is nasty, but it gives it windows support
		else if( this.isWindows() && signal == 'SIGTERM' ) signal = 'SIGUSR2';
		// asked to restart
		else if( code == 2 ) signal = 'SIGUSR2';

		this.process = null;
		this.state = 'killed';

		// exit the monitor, but do it gracefully
		if( signal == 'SIGUSR2' ){
			this.restart();
		}
		// clean exit - wait until file change to restart
		else if( code === 0 ){
			if( this.hasRestartFile ){
				this.console.log('nodeprocess stopped - {path}', this);
			}
			this.emit('stop');
		}
		else{
			if( this.hasRestartFile ){
				this.console.error('nodeprocess crashed {path}', this);
			}
			this.emit('crash');
		}
	},

	send: function(message, handle){
		this.process.send(message, handle);
	},

	onmessage: function(message, handle){
		if( message.type ){
			this.emit(message.type, message.event);
		}
	}
});

module.exports = NodeProcess;
