// unitTest/TestSerie/index.js
/*

*/

var TestPrototype = require('../TestPrototype');

var TestSerie = TestPrototype.create({
	Test: require('../Test'),
	name: 'Anonymous testSerie',
	tests: null,

	init: function(tests){
		this.tests = tests;
	},
	
	setModule: function(module){
		this.tests.forEach(function(test){
			test.module = module;
		});
		this.module = module;
	},

	createTest: function(){
		return this.Test.new.apply(this.Test, arguments);
	},

	add: function(){
		var test = this.createTest.apply(this, arguments);
		this.tests.push(test);
		return this;
	},

	forEach: function(fn, bind){
		return this.tests.forEach(fn, bind);
	},

	setup: function(){
		this.index = 0;
		this.current = null;
	},

	teardown: function(){
		// something to do?
	},

	clean: function(){
		if( this.current ){
			this.current.stop();
		}
	},

	next: function(){
		if( this.index >= this.tests.length ){
			return null;
		}
		else{
			this.current = this.tests[this.index];
			this.index++;
			return this.current;
		}
	},

	nextTest: function(e){
		if( e ){
			/*
			var stackTrace = require('stackTrace');
			var call = stackTrace.findCallByFile(stackTrace.get(), this.testSerie.path);
			var line = call.getLineNumber();
			
						
			error.file = this.testSerie.path;
			error.line = line - 1;
			*/
			this.fail(e);
		}
		else if( this.next() ){
			this.current.start(this.nextTest, this);
		}
		// all test passed with success or no test to run
		else{
			this.pass();
		}
	},

	run: function(){
		this.nextTest();
	}
});

module.exports = TestSerie;
