// Router/middleware/file/File/index.js
var proto = require('proto');
var Path = require('path');
require('promise-extra');
var File = proto.create({
	fileSystem: require('fs'),
	path: null,
	modifiedSince: null,
	acceptGzip: null,
	stat: null,
	aliases: [],
	aliasIndex: 0,
	alias: null,

	constructor: function(path){
		this.path = this.originalPath = Path.normalize(path);
		this.result = {};
	},

	isModified: function(mtime){
		return this.modifiedSince ? mtime > this.modifiedSince : true;
	},

	isNotFoundError: function(error){
		return error.code == 'ENOENT' && error.errno == 34;
	},

	getStat: function(){
		return Promise.from(function(complete){
			this.fileSystem.stat(this.path, complete);
		}, this);
	},

	nextAlias: function(){
		var aliases = this.aliases;

		if( aliases.length === 0 ) return null;
		if( this.aliasIndex >= aliases.length ) return null;
		return aliases[this.aliasIndex++].new(this);
	},

	nextPossibleAlias: function(){
		var alias;

		while( (alias = this.nextAlias()) && alias.handle() );

		return alias;
	},

	sendAliasElseFile: function(){
		this.path = this.originalPath;
		this.alias = this.nextPossibleAlias();

		var promise = this.getStat();

		if( this.alias ){
			this.path = this.alias.getPath(this.path);

			promise = promise.catch(function(error){
				if( this.isNotFoundError(error) ){
					this.path = this.originalPath;
					return this.sendAliasElseFile();
				}
				return Promise.reject(error);
			}.bind(this));
		}

		return promise.catch(function(error){
			if( this.isNotFoundError(error) ) return 404;
			return Promise.reject(error);
		}.bind(this)).then(function(stat){
			// seul les fichiers sont autorisé
			if( !stat.isFile() ) return 403;
			// dit au navigateur que le fichier n'a pas changé
			if( !this.isModified(stat.mtime) ) return 304;

			return {
				statusCode: 200,
				headers: {
					'content-type': config.getMimetype(this.originalPath),
					'last-modified': stat.mtime,
					'content-length': stat.size,
					// évite que chrome mette en cache et réutilise sans redemander au serveur
					// les fichier HTML qu'on lui envoit
					'cache-control': 'no-cache'
				},
				body: require('fs').createReadStream(this.path), //this.fileSystem.createReadStream(this.path),
				// dont log body for files
				logBody: !false
			};
		}.bind(this));
	},

	generate: function(){
		return this.sendAliasElseFile();
	}
});

module.exports = File;
