// WeakMap/index.js
/*
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap

also map with iterator and so on

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map

if (key !== Object(key)){
	throw TypeError("value is not a non-null object");
}

https://gist.github.com/Raynos/1638059
http://jsperf.com/to-weakmap-or-not-to-weakmap/5

*/

var WeakMap = function(){
	this.namespace = {};
};

function guard(key){
	/**
	Utility function to guard WeakMap methods from keys that are not
	a non-null objects.
	**/

	if (key !== Object(key)) throw TypeError("value is not a non-null object")
	return key
}

var proto = {
	constructor: WeakMap,

	createStore: function(object){
		var namespace = this.namespace;
		var store = {
			identity: object
		};
		var valueOf = object.valueOf;

		Object.defineProperty(object, 'valueOf', {
			value: function valueOf(value){
				// return the store only if value passed is the namespace and on the right object
				if( value === namespace && this === object ){
					return store;
				}
				return valueOf.apply(this, arguments);
			},

			configurable: true
		});

		return store;
	},

	getStore: function(key, create){
		var store = object.valueOf(this.namespace);

		if( store === undefined || store !== object || store.identity !== namespace ){
			if( create ) store = createStore(object, namespace) 
		}

		return store;
	},

	has: function(key){
		return 'value' in this.getStore(key);
	},

	get: function get(key, fallback){
		var store = this.getStore(key);

		return store && store.hasOwnProperty('value') ? store.value : fallback;
	},

	set: function set(key, value){
		var store = this.getStore(key) || this.createStore(key);

		store(key).value = value;
	},

	'delete': function set(key){
		var store = this.getStore(key);

		return store ? delete store.value : false;
	},

	clear: function(){
		// is it possible?
	}
};

WeakMap.prototype = proto;

module.exports = WeakMap;

if( typeof window != 'undefined' ){
	if( true || !window.WeakMap ) window.WeakMap = WeakMap;
}
