// StringTemplate/index.js
/*



*/

var proto = require('proto');

var ExpressionsResolver = {
	rawExpression: function(expression){
		return '{' + expression + '}';
	},

	getExpressionValueInNull: function(expression){
		return this.rawExpression(expression);
	},

	getExpressionValueInArray: function(expression, array, index){
		if( index in array ) return array[index];
		return this.rawExpression(expression);
	},

	getExpressionValueInObject: function(expression, object, index){
		// replace 'name' by object.name
		if( expression in object ) return object[expression];
		// replace 'name' by '{name}'
		return this.rawExpression(expression);
	},

	resolve: function(expressions, scope, map, bind){
		var i = 0, j = expressions.length, values = new Array(j), value, getter;

		if( scope == null ){
			getter = this.getExpressionValueInNull;
		}
		else if( scope instanceof Array ){
			getter = this.getExpressionValueInArray;
		}
		else{
			getter = this.getExpressionValueInObject;
		}

		for(;i<j;i++){
			value = getter.call(this, expressions[i], scope, i);
			if( map ) value = map.call(bind, value, i, scope);
			values[i] = value;
		}

		return values;
	}
};

var Renderer = {
	reduce: function(strings, values, fn, bind, buffer){
		var i = 0, j = values.length, string, value;

		while(i<j){
			value = String(values[i]);
			string = strings[i + 1];
			buffer = fn.call(bind, buffer, value, string, i);
			i++;
		}

		return buffer;
	},

	stringReducer: function(prev, value, string){
		return prev + value + string;
	},

	arrayReducer: function(prev, value, string){
		prev.push(value, string);
		return prev;
	},

	renderAsString: function(strings, values){
		return this.reduce(strings, values, this.stringReducer, this, strings[0]);
	},

	renderAsArray: function(strings, values){
		return this.reduce(strings, values, this.arrayReducer, this, [strings[0]]);
	}
};

var StringTemplate = proto.create({
	Compiler: require('./Compiler'),
	Resolver: ExpressionsResolver,
	Renderer: Renderer,

	_string: null,
	_strings: null,
	_expressions: null,
	_scope: null,
	_values: null,

	init: function(string){
		this.string = string;		
	},

	compile: function(string){
		return this.Compiler.compile(string);
	},

	resolve: function(scope, map, bind){
		return this.Resolver.resolve(this.expressions, scope, map, bind);
	},

	set string(value){
		this._string = value;
		this.strings = null;
		this.expressions = null;
		this.values = null;
	},

	get string(){
		return this._string;
	},

	set strings(value){
		this._strings = value;
	},

	set expressions(value){
		this._expressions = value;
	},

	get strings(){
		if( this._strings === null ) this.result = this.compile(this.string);
		return this._strings;
	},

	get expressions(){
		if( this._expressions === null ) this.result = this.compile(this.string);
		return this._expressions;
	},

	set result(value){
		this.strings = value.strings;
		this.expressions = value.expressions;
	},

	set scope(value){
		this._scope = value;
		this.values = null;
	},

	get scope(){
		return this._scope;
	},

	set values(values){
		this._values = values;
	},

	get values(){
		var value = this._values;
		if( value == null ){
			value = this.resolve(this.scope);
			this._values = value;
		}
		return value;
	},

	toString: function(){
		return this.Renderer.renderAsString(this.strings, this.values);
	},

	toArray: function(){
		return this.Renderer.renderAsString(this.strings, this.values);
	},

	exec: function(scope, map, bind){
		this.scope = scope;
		this.values = this.resolve(scope, map, bind);
		return this;
	},

	render: function(scope, map, bind){
		this.exec(scope, map, bind);
		return this.toString();
	},

	tag: function(fn, bind){
		return fn.apply(bind, [this.strings].concat(this.values));
	}
});

/*
String.format = function(string){
	var tpl = StringTemplate.new(string);

	if( arguments.length > 1 ){
		var scope = Array.prototype.slice.call(arguments, tpl.expressions.length);
		tpl.exec(scope);
	}

	return tpl.toString();
};
*/

module.exports = StringTemplate;
