// StringTemplate/index.js

require('Object.instance');

var Splitter = require('Splitter');
var TagSplitter = Splitter.extend({
	cache: {},
	open: '{',
	close: '}',
	regexp: null,

	init: function(){
		Splitter.init.apply(this, arguments);
		this.regexp = new RegExp(this.open + '(.*?)' + this.close);
	},

	_split: function(string){
		return string.split(this.regexp);
	}
});

var StringTemplate = {
	Parser: TagSplitter,
	tokens: null,

	init: function(string){
		this.parser = this.Parser.new(string);
	},

	get parts(){
		if( this.hasOwnProperty('tokens') === false ){
			this.tokens = [].concat(this.split());
		}
		return this.tokens;
	},

	parse: function(){
		return this.parser.split();
	},

	empty: function(token){
		return this.parser.open + token + this.parser.close;
	},

	replacer: function(token, data){
		if( token in data ) return data[token];
		return this.empty(token);
	},

	stringify: function(value){
		return value;
	},

	// replace one token variable part at index with value
	replaceOne: function(tokens, index, value){
		value = this.stringify(value);
		tokens[index] = value;
		return tokens.join();
	},

	// replace all variable tokens with value found in data
	replace: function(tokens, data, replacer, bind){
		if( typeof replacer != 'function' ) throw new TypeError('replacer must be a function');

		var i, j, isVariable = false, value;

		result = [];
		i = 0;
		j = tokens.length;
		
		while( i < j ){
			if( isVariable === true ){
				value = replacer.call(bind, tokens[i], data, i);
				value = this.stringify(value);
				tokens[i] = value;
				isVariable = false;
			}
			else{
				isVariable = true;
			}
			i++;
		}

		return tokens.join();
	},

	render: function(data){
		return this.replace(this.parts, data, this.replacer, this);
	}
};

module.exports = StringTemplate;
