// StringTemplate/StringTemplate.js
/*

*/

var proto = require('proto');

var StringTemplate = proto.create({
	Compiler: require('./Compiler'),
	Resolver: require('./Resolver'),
	Renderer: require('./Renderer'),
	string: null,
	scope: null,
	map: null,
	bind: null,
	mergePropertyNames: ['string', 'compilationResult', 'values'],
	merge: require('property/merge').mergeProperties,

	init: function(string){
		if( this.isPrototypeOf(string) ){
			this.populate(string);
		}
		else{
			this.string = string;
		}
	},

	populate: function(template){
		this.merge(this, template, this.mergePropertyNames);
		return template;
	},

	compile: function(string){
		return this.Compiler.compile(string);
	},

	get compilationResult(){
		return this.compile(this.string);
	},

	set strings(value){
		this.compilationResult.strings = value;
	},

	set expressions(value){
		this.compilationResult.expressions = value;
	},

	get strings(){
		return this.compilationResult.strings;
	},

	get expressions(){
		return this.compilationResult.expressions;
	},

	get values(){
		return this.Resolver.resolve(this.expressions, this.scope, this.map, this.bind);
	},

	toString: function(){
		return this.Renderer.renderAsString(this.strings, this.values);
	},

	toArray: function(){
		return this.Renderer.renderAsArray(this.strings, this.values);
	},

	tag: function(fn, bind){
		return fn.apply(bind, [this.strings].concat(this.values));
	},

	exec: function(scope, map, bind){
		this.scope = scope;
		this.map = map;
		this.bind = bind;
		this.values = null;
		return this;
	},

	render: function(scope, map, bind){
		return this.exec(scope, map, bind).toString();
	}
});

/*
template.compilationResult value is cached
To invalidate the cache set the property to null or undefined
template.compilationResult = null
To set the cache without calling the getter, set the property to a value not null nor undefined
template.compilationResult = {string: [], expressions: []};
*/

StringTemplate.cacheGetter = function(propertyName){
	var descriptor = Object.getOwnPropertyDescriptor(this, propertyName);
	var getter = descriptor.get;
	var hasProperty = false, propertyValue;

	Object.defineProperty(this, propertyName, {
		get: function(){
			if( hasProperty === false ){
				hasProperty = true;
				propertyValue = getter.call(this);
			}

			return propertyValue;
		},

		set: function(value){
			if( value == null ){
				hasProperty = false;
			}
			else{
				hasProperty = true;
				propertyValue = value;
			}			
		}
	});
};
StringTemplate.cacheGetter('compilationResult');
StringTemplate.cacheGetter('values');

/*
String.format = function(string){
	var tpl = StringTemplate.new(string);

	if( arguments.length > 1 ){
		var scope = Array.prototype.slice.call(arguments, tpl.expressions.length);
		tpl.exec(scope);
	}

	return tpl.toString();
};
*/

module.exports = StringTemplate;
