// Router/middleware/module/RequireContext/index.js
/*

*/

var proto = require('proto');
var Resolver = require('../RequireResolver');
require('String.prototype.startsWith');

var RequireContext = proto.create({
	Resolver: Resolver,
	clientFolder: null,
	cache: {},
	resolvedPaths : null,
	meta: null, // data from package.json file
	rootFolder: process.cwd(),

	constructor: function(path, parent){
		//var filename = require('path').resolve(path, this.rootFolder);
		var filename = path;

		if( filename in this.cache ){
			return this.cache[filename];
		}
		
		this.cache[filename] = this;
		this.filename = filename;
		this.parent = parent;
		this.resolvedPaths = {};
		this.children = [];

		if( parent ){
			parent.children.push(module);
		}
	},

	createChild: function(filename){
		return new RequireContext(filename, this);
	},

	isClientReadable: function(){
		if( this.clientFolder && this.filename.startsWith(this.clientFolder) ) return true;
		if( this.meta && this.meta.clientReadable === true ) return true;
		return false;
	},

	resolve: function(path){
		if( path in this.resolvedPaths ) return Promise.resolve(this.resolvedPaths[path]);

		var resolver = new Resolver(this.filename, path, this.rootFolder);
		var p = resolver.resolve();

		return p.then(function(resolvedPath){
			this.meta = resolver.packageBody;
			// make filename relative to the root path
			var filename = resolvedPath;//resolver.relativePath(resolvedPath);

			this.resolvedPaths[path] = filename;
			this.createChild(filename);
			return resolvedPath;
		}.bind(this));
	},

	toJSON: function(){
		return {
			filename: this.filename,
			resolvedPaths: this.resolvedPaths,
			children: this.children
		};
	}
});

module.exports = RequireContext;
