// TokenList/index.js
/*

name: TokenList

description: TokenList is used to manipulate a list of string separated by a space (similar to native classList)

https://developer.mozilla.org/en-US/docs/Web/API/Element.classList

*/

require('core/string'); // for trim()
require('Object.instance');

var List = Array.prototype.extend({
	init: function(){
		if( arguments.length > 0 ) this.push.apply(this, arguments);
	},

	clear: function(){
		this.length = 0;
	}
});

var TokenList = List.extend({
	init: function(string){
		if( string ){
			var trimmed = String.prototype.trim.call(string);
			if( trimmed ) List.init.apply(this, trimmed.split(/\s+/));
		}
		return this;
	},

	indexOf: function(token, index){
		if( token === '' ) throw new Error('SYNTAX_ERR', 'An invalid or illegal string was specified');
		if( /\s/.test(token) ) throw new Error('INVALID_CHARACTER_ERR', 'String contains an invalid character');

		return Array.prototype.indexOf.call(this, token, index);
	},

	contains: function(token, index){
		return this.indexOf(token, index) !== -1;
	},

	update: function(){

	},

	add: function(){
		var tokens = arguments, i = 0, j = tokens.length, token, updated = false;

		for(;i<j;i++){
			token = String(tokens[i]);
			if( !this.contains(token) ){
				this.push(token);
				updated = true;
			}
		}
		if( updated ) this.update();

		return this;
	},

	remove: function(){
		var tokens = arguments, i = 0, j = tokens.length, token, updated = false, index;

		for(;i<j;i++){
			token = String(tokens[i]);
			index = this.indexOf(token);
			if( index !== -1 ){
				this.splice(index, 1);
				updated = true;
			}
		}
		if( updated ) this.update();

		return this;
	},

	toggle: function(token, force){
		if( arguments.length === 1 ) force = !this.contains(token);
		return force ? this.add(token) : this.remove(token);
	},

	clear: function(){
		List.clear.call(this);
		return this.update();
	},

	toString: function(){
		return this.join(' ');
	}
});

module.exports = TokenList;
