// PropertyAccessor/Splitter/index.js
/*



*/

var ParentSplitter = require('Splitter');
var Splitter = ParentSplitter.extend({
	Part: require('../Part'),
	specials: '.[]()',
	singleQuote: '\'',
	doubleQuote: '"',
	
	isSpecial: function(char){
		return this.specials.indexOf(char) !== -1;
	},

	isNumber: function(char){
		return '0' <= char && char <= '9';
	},

	isAlpha: function(char){
		return (char >= 'a' && char <= 'z') || (char >= 'A' && char <= 'Z') || char === '_' || char === '$';
	},

	isAlphaNum: function(char){
		return this.isAlpha(char) || this.isNumber(char);
	},

	isSingleQuote: function(char){
		return char == this.singleQuote;
	},

	isDoubleQuote: function(char){
		return char == this.doubleQuote;
	},

	unexpectedChar: function(){
		throw this.createError('Unexpected char ' + this.char + ' at ' + this.index + ' in ' + this.buffer);
	},

	unterminatedQuote: function(index){
		throw this.createError('Unterminated quote at ' + index + ' in ' + this.buffer);
	},

	createPartFromSlice: function(){
		var part = ParentSplitter.createPartFromSlice.apply(Splitter, arguments);

		if( this.char == '"' || this.char == '\'' ) this.index++;
		if( this.char == '(' && this.next() && this.char == ')' ){
			// token.args = eval('['+ this.buffer.slice(this.index, this.getSeparatorIndex(')'))  +']''); 
			// et faire pareil pour les expression entre []
			// tention: gros problème de sécurité si j'autorise ça coté serveur
			part.args = [];
		}

		return part;
	},

	createNextPart: function(){
		var start, end;

		while( this.next() ){
			if( this.isSpecial(this.char) ){
				continue;
			}

			// begining of a token
			if( this.isAlphaNum(this.char) ){
				start = this.index;
				while( this.next() && this.isAlphaNum(this.char) );
				end = this.index;

				break;
			}

			if( this.isSingleQuote(this.char) ){
				start = this.index;
				do{
					if( this.next() === false ){
						this.unterminatedQuote(start);
					}
				}
				while( this.isSingleQuote(this.char) === false );
				end = this.index;

				break;	
			}

			if( this.isDoubleQuote(this.char) ){
				start = this.index;				
				do{
					if( this.next() === false ){
						this.unterminatedQuote(start);
					}
				}
				while( this.isDoubleQuote(this.char) === false );
				end = this.index;

				break;
			}

			this.unexpectedChar();
		}

		return this.createPartFromSlice(start, end);
	}
});

module.exports = Splitter;
