// Router/middleware/module/Module/index.js
var proto = require('proto');
var Module = proto.extend({
	ModuleResolver: require('../ModuleResolver'),
	Path: require('path'),

	cache: {},
	resolvedPaths : null,
	new: function(filename, parent){
		// get filename using clientPath
		filename = this.Path.resolve(global.CLIENT_PATH, filename);

		var module;

		if( filename in this.cache ){
			module = this.cache[filename];
		}
		else{
			module = proto.new.call(this, filename, parent);
			this.cache[filename] = module;
		}

		if( parent ){
			parent.children.push(module);
		}

		return module;
	},

	init: function(filename, parent){
		this.filename = filename;
		this.parent = parent;
		this.resolvedPaths = {};
		this.children = [];
	},

	createChild: function(filename){
		return Module.new(filename, this);
	},

	resolve: function(path, fn, bind){
		if( path in this.resolvedPaths ){
			process.nextTick(function(){
				fn.call(bind, null, this.resolvedPaths[path]);
			}.bind(this));
		}
		else{
			var resolver = this.ModuleResolver.new(this.filename, path);

			resolver.resolve(function(error, filename){
				if( error ){
					fn.call(bind, error);
				}
				else if( filename ){
					// make filename relative to the client path
					filename = this.Path.relative(global.CLIENT_PATH, filename);

					this.resolvedPaths[path] = filename;
					this.createChild(filename);
					fn.call(bind, filename);
				}
				else{					
					fn.call(bind, new Error('not found'));
				}
			}.bind(this));
		}
	},

	toJSON: function(){
		return {
			filename: this.filename,
			resolvedPaths: this.resolvedPaths,
			children: this.children
		};
	}
});

module.exports = Module;
