// ArrayIterator/index.js
var proto = require('proto');
var ArrayIterator = proto.create({
	loopIndex: null,
	looped: false,
	array: null,
	direction: 'next',	
	from: null,
	to: null,
	sign: null,	
	start: null,
	end: null,
	done: {done: true},
	index: null,
	length: null,

	init: function(array, direction, from, to){
		if( array == null ){
			throw new TypeError("array is null or not defined");
		}

		direction = direction || this.direction;

		var length, sign, start, end;

		length = Object(array).length >>> 0;

		if( direction === 'next' ){
			sign = 1;
			start = -1;
			end = length - 1;
		}
		else{
			if( direction === 'both' ) to = true;
			sign = -1;
			start = length;
			end = 0;
		}

		//if( length === 0 ) return;

		// from should respect array limits
		if( typeof from != 'number' ) from = start;
		else if( sign == 1 ? from < start : from > start ) from = start;
		else if( sign == 1 ? from > end : from < end ) from = end;

		// if to === true we loop twice except if from == start as loop is useless
		if( to === true && from != start ) this.loopIndex = from;

		// to should respect array limits
		if( typeof to != 'number' ) to = end;
		else if( sign == 1 ? to > end : to < end ) to = end;
		else if( sign == 1 ? to < start : to > start ) to = start;

		// prevent infinite loop
		if( sign == 1 ? to < from : to > from ){
			to = from;
			this.loopIndex = null;
		}

		this.array = array;
		this.direction = direction;
		this.from = from;
		this.to = to;

		this.index = this.from;
		this.length = this.to;
	},

	next: function(){

		if( this.index === this.length ){
			if( this.looped === true || typeof this.loopIndex != 'number' ) return this.done;
			this.index = this.start;
			this.length = this.loopIndex;
			this.looped = true;
		}

		this.index+= this.sign;
		return {value: this.array[this.index], done: false};
	},

	/*
	reset: function(){
		this.index = this.from;
		this.length = this.to;
		this.looped = false;
	},
	*/

	match: function(fn, bind){
		var next;

		while( true ){
			next = this.next();
			if( next.done === true ) break;
			if( !fn || fn.call(bind, next.value, this.index, this.array) === true ) return next.value;
		}

		return null;
	},

	matchAll: function(fn, bind){
		var matches = [], next;

		while( true ){
			next = this.next();
			if( next.done === true ) break;
			if( !fn || fn.call(bind, next.value, this.index, this.array) === true ) matches.push(next.value);
		}

		return matches;
	}
});

module.exports = ArrayIterator;
