// Object.property/index.js
/*

name: Object.property

description: required almost everywhere

provides:
	ownKeys, eachOwnKey, iterateKey, eachOwnKeyIn
	appendKey, append, completeKey, complete
	implementKey, implement, complementKey, complement

*/

module.exports = {
	ownKeys: 'getOwnPropertyNames' in Object ? Object.getOwnPropertyNames : Object.keys,

	eachOwnKey: function(object, fn, bind){
		var keys = this.ownKeys(object), i = 0, j = keys.length, key;

		for(;i<j;i++){
			key = keys[i];
			fn.call(bind, key, object);
		}

		return object;
	},

	iterateKey: function(keyIterator, array, fn, bind){
		var i = 0, j = array.length, item;

		for(;i<j;i++){
			item = array[i];
			if( item instanceof Function ) item = item.prototype;

			switch(typeof item){
			case 'string':
				var temp = {};
				temp[item] = array[++i];

				fn.call(bind, item, temp);
				break;
			case 'object':
				keyIterator.call(this, item, fn, bind);
				break;
			}
		}
	},

	eachOwnKeyIn: function(array, fn, bind){
		return this.iterateKey(this.eachOwnKey, array, fn, bind);
	},

	append: function(object){
		this.eachOwnKeyIn(Array.prototype.slice.call(arguments, 1), this.appendKey, object);
		return object;
	},

	complete: function(object){
		this.eachOwnKeyIn(Array.prototype.slice.call(arguments, 1), this.completeKey, object);
		return object;
	},

	implement: function(object){
		this.eachOwnKeyIn(Array.prototype.slice.call(arguments, 1), this.implementKey, object.prototype);
		return this;
	},

	complement: function(object){
		this.eachOwnKeyIn(Array.prototype.slice.call(arguments, 1), this.complementKey, object.prototype);
		return this;
	}
};

// avoir custom getter being called
module.exports.cloneDescriptor = function(descriptor){
	var clone = {}, keys, i, j, key;

	keys = module.exports.ownKeys(descriptor);
	i = 0;
	j = keys.length;

	for(;i<j;i++){
		key = keys[i];
		clone[key] = descriptor[key];
	}

	if( !descriptor.set ) delete descriptor.set;

	return clone;
};

if( 'getOwnPropertyDescriptor' in Object ){

	module.exports.appendKey = function(key, object){
		if( object ){
			var descriptor = Object.getOwnPropertyDescriptor(object, key);
			// not needed as we want the same descriptor as source
			//descriptor = module.exports.cloneDescriptor(descriptor);
			Object.defineProperty(this, key, descriptor);
		}
		else{
			this[key] = object[key];
		}
	};

	module.exports.implementKey = function(key, object){
		var descriptor;

		if( object ){
			descriptor = Object.getOwnPropertyDescriptor(object, key);
			// don't put enumerable = false on source descriptor
			descriptor = module.exports.cloneDescriptor(descriptor);
			descriptor.enumerable = false;
		}
		else{
			descriptor = {enumerable: false,  configurable: true, writable: true, value: object[key]};
		}

		Object.defineProperty(this, key, descriptor);
	};

}
else{

	module.exports.appendKey = module.exports.implementKey = function(key, object){
		this[key] = object[key];
	};

}

module.exports.completeKey = function(key){
	// set key/value pair in this if not existing
	if( !(key in this) ){
		module.exports.appendKey.apply(this, arguments);
	}
	return this;
};

module.exports.complementKey = function(key){
	if( !(key in this) ){
		module.exports.implementKey.apply(this, arguments);
	}
	return this;
};

Object.append = module.exports.append.bind(module.exports);
