// unitTest/AppTest/ModuleLoader/index.js
/*
*/

var ModuleLoader = {
	fileSystem: require('fs'),
	path: require('path'),
	moduleFolderName: 'node_modules',

	hasModuleSignature: function(path){
		// one of his parent is node_modules but the fileName != 'node_modules'
		var directories = path.split(this.path.sep);
		var index = directories.indexOf(this.moduleFolderName);

		return index > -1 && index < directories.length - 1;
	},

	filterFile: function(fileName){
		return fileName == '.git' || fileName == '.gitignore' || fileName == '.gitattributes';
	},

	load: function(path, badSignature, ignoreFile){
		var fileNames, i, j, fileName, filePath, modulePaths, hasModuleSignature, modulePath;

		modulePaths = [];
		modulePath = null;
		fileNames = this.fileSystem.readdirSync(path);
		i = 0;
		j = fileNames.length;
		for(;i<j;i++){
			fileName = fileNames[i];
			if( this.filterFile(fileName) ) continue;

			filePath = path + this.path.sep + fileName;			

			if( this.fileSystem.statSync(filePath).isDirectory() ){
				hasModuleSignature = this.hasModuleSignature(filePath);

				if( hasModuleSignature ){
					try{
						// locate module path for this folder
						modulePath = require.resolve(filePath);
					}
					catch(e){
						if( badSignature ){
							badSignature.push(filePath);							
						}
					}

					if( modulePath ) modulePaths.push(modulePath);
				}

				modulePaths = modulePaths.concat(this.find(filePath, badSignature, modulePath != null));
				modulePath = null;
			}
			/*
			else if( !ignoreFile && this.hasModuleSignature(filePath) ){
				modulePaths.push(filePath);
			}
			*/
		}

		return modulePaths;
	}
};

module.exports = ModuleLoader;
