// StringTemplate/StylizedTemplate/index.js
/*

StylizedTemplate.registerFormat('background-yellow', {background: 'yellow'});
StylizedTemplate.registerFormat('red', {color: 'red'});
StylizedTemplate.registerFormat('green', {color: 'green'});

var template = StylizedTemplate.new('Hello {name}, age: {age} !').registerStyles(['green', 'red']);

template.render({name: 'damien', age: 10});

var template = StylizedTemplate.new('Hello {name} !').registerStyles({name: 'green'});
template.globalStyle = 'background-yellow';
template.render({name: 'damien', age: 10});

*/

var proto = require('proto');
var StringTemplate = require('StringTemplate');

var TemplateResult = StringTemplate.Result;
var StylizedTemplateResult = TemplateResult.create({
	formats: null,
	styles: null,
	globalStyle: null,
	prefix: null,
	suffix: null,

	init: function(template){
		TemplateResult.init.call(this, template);

		this.globalStyle = template.globalStyle;
		this.formats = Object.create(template.formats || null);
		this.styles = Object.create(template.styles || null);

		var key;
		for(key in template.customFormats) this.formats[key] = template.customFormats[key];
		for(key in template.customStyles) this.styles[key] = template.customStyles[key];

		this.prefix = template.prefix;
		this.suffix = template.suffix;
	},

	writeStyle: function(string, style){
		throw new Error('unimplemented method');
	},

	stylizeValue: function(value, styleName){
		
		// use StylizedTemplate & StylizedTemplateResult stylizedString representation
		if( typeof value == 'object' && typeof value['toStylizedString'] == 'function' ){
			value = value.toStylizedString();
		}
		if( styleName in this.styles ){
			value = String(value);
			value = this.writeStyle(value, this.styles[styleName]);
		}

		return value;
	},

	stylizeValueAt: function(value, index){
		return this.stylizeValue(value, this.expressions[index]);
	},

	stylizeValues: function(values){
		var i = 0, j = values.length, stylizedValues = new Array(j);

		for(;i<j;i++){
			stylizedValues[i] = this.stylizeValueAt(values[i], i);
		}

		return stylizedValues;
	},

	stringify: function(object, stylized){
		return stylized && object.toStylizedString == 'function' ? object.toStylizedString() : String(object); 
	},

	toString: function(){
		var string;

		string = TemplateResult.toString.call(this);
		if( this.prefix ) string = this.prefix + string;
		if( this.suffix ) string+= this.suffix;

		return string;
	},

	toStylizedString: function(){
		var string;

		string = this.Renderer.renderAsString(this.strings, this.stylizeValues(this.values));
		if( this.prefix ) string = this.prefix + string;
		if( this.suffix ) string+= this.suffix;
		string = this.stylizeValue(string, this.globalStyle);

		return string;
	}
});

var StylizedTemplate = StringTemplate.create({
	Result: StylizedTemplateResult,
	formats: {},
	styles: {},
	customFormats: null,
	customStyles: null,
	globalStyle: null,
	prefix: null,
	suffix: null,

	init: function(string){
		StringTemplate.init.call(this, string);

		this.customFormats = {};
		this.customStyles = {};
	},

	// format are used to represent a list of styles by a string
	registerFormat: function(name, styles){
		if( typeof styles == 'string' ) styles = [styles];

		// when called from instance we add it to customFormats
		this[this.hasOwnProperty('string') ? 'customFormats' : 'format'][name] = styles;
	},

	// style are used to bind a format to an expression
	registerStyle: function(name, format){
		if( typeof format != 'string' ) this.registerFormat(name, format);
		
		this[this.hasOwnProperty('string') ? 'customStyles' : 'style'][name] = format;
	},
	
	registerFormats: function(formats){
		for(var format in formats){
			this.registerFormat(format, formats[format]);
		}
	},

	registerStyles: function(styles){
		var key;

		if( styles instanceof Array ){
			var expressions = this.expressions;
			for(key in styles){
				this.registerStyle(expressions[key], arguments[key]);
			}
		}
		else{
			for(key in styles){
				this.registerStyle(key, styles[key]);
			}
		}
	},

	toStylizedString: function(){
		return this.result.toStylizedString();
	}
});

module.exports = StylizedTemplate;
