// router/middleware/bodyReader/header/requestHeader.js
var RequestHeader = {
	name: null,
	value: null,
	bodyReader: null,
	request: null,
	error: null,
	errorMessage: 'Invalid value {headerValue} for request header {headerName}.',
	errorStatusCode: 400, //  bad request
	Error: require('headerValueError'),

	new: function(bodyReader){
		if( this.value !== null || this.name in bodyReader.request.headers || 'getValue' in this ){
			return Object.prototype.new.call(this, bodyReader);
		}
		else{
			return false;
		}
	},

	create: function(bodyReader){
		this.bodyReader = bodyReader;
		this.request = bodyReader.request;

		var value;
		if( this.name in this.request.headers ){
			value = this.request.headers[this.name];
		}
		else if( this.value === null ){
			value = this.getValue();
		}
		else{
			value = this.value;
		}

		if( this.parse ) value = this.parse(value);
		this.value = value;
	},

	createError: function(statusCode, message, data){
		statusCode = statusCode || this.errorStatusCode;
		message = message || this.errorMessage;
		data = data || {};

		if( !('headerName' in data) ) data.headerName = this.name;
		if( !('headerValue' in data) ) data.headerValue = this.value;

		var error = new this.Error(statusCode, message, data);
		return error;
	},

	isValid: function(){
		return true;
	},

	valid: function(){
		if( this.isValid() ){
			return true;
		}
		this.error = this.createError();
		return false;
	},

	effect: function(value){
		// noop
	}
};

module.exports = RequestHeader;
