// mdv/Template/TemplateCompiler/TokenListLinker/index.js
/*

name: TokenListLinker
description: link a node attribute to a model multiple values (with prefix and suffix)
example: <span>Hello {name}, you are {age} years old!</span>

*/
require('Node.bind');
var Linker = require('./linker');

var TokenListLinker = Linker.create({
	ComputedObserver: require('ComputedObserver'),
	name: 'TokenListLinker',
	nodeAttribute: null,
	stringTemplate: null,

	init: function(nodeAttribute, stringTemplate){
		this.nodeAttribute = nodeAttribute;
		this.stringTemplate = stringTemplate;
	},

	replaceAll: function(values){
		return this.StringTemplate.render(values);
	},

	replaceOne: function(values, observer){
		this.StringTemplate.values[observer.id] = observer.lastChange.value;
		return this.StringTemplate.toString();
	},

	link: function(node, model){
		var expressions = this.stringTemplate.expressions, i = 0, j = expressions.length, observer;

		observer = this.ComputedObserver.new();
		for(;i<j;i++){
			// observe part, and pass true to avoid resolve call for each observer call
			observer.observe(i, model, expressions[i], true);
		}
		observer.value = this.replaceAll(observer.values);
		observer.combinator = this.replaceOne;
		observer.combinatorBind = this;

		node.bind(this.nodeAttribute, observer, 'value');
	},

	unlink: function(node, model){
		node.unbind(this.nodeAttribute);
	},

	namedScope: function(path, replace){
		var i = 1, expressions = this.stringTemplate.expressions, j = expressions.length;
		for(;i<j;i+=2){
			expressions[i] = this.getNamedScopePath(expressions[i], path, replace);
		}
	}
});

module.exports = TokenListLinker;
