// proto/factory/index.js
var proto = require('proto');

var ProtoFactory = proto.extend({
	plugins: [],
	index: -1,
	effects: {},

	constructor: function(plugins){
		this.plugins = [];

		if( plugins ){
			var i = 0, j = plugins.length;
			for(;i<j;i++){
				this.add(plugins[i]);
			}
		}
	},

	add: function(plugin){
		if( Object(plugin) != plugin ){
			throw new TypeError('plugin must be an object');
		}

		if( !this.has(plugin) ){
			this.plugins.push(plugin);
		}
	},

	has: function(plugin){
		this.index = this.plugins.indexOf(plugin);
		return this.index !== -1;
	},

	remove: function(plugin){
		if( this.has(plugin) ){
			this.plugins.splice(this.index, 1);
		}
	},

	productConstructor: function(){
		// called with a created product as context
	},

	createRawProduct: function(object){
		var constructors = this.plugins.filter(function(plugin){
			return plugin.hasOwnProperty('constructor');
		}).map(function(plugin){
			return plugin.constructor;
		});

		var rawProduct = proto.extend.call(object, {
			constructor: function(){
				var args = arguments;
				constructors.forEach(function(constructor){
					constructor.apply(this, args);
				}, this);
				return object.constructor.apply(this, arguments);
			}
		});

		this.productConstructor.call(rawProduct);

		return rawProduct;
	},

	refine: function(rawProduct){
		this.plugins.forEach(function(plugin){
			for(var key in this.effects){
				if( plugin.hasOwnProperty(key) ){
					this.effects[key].call(this, rawProduct, plugin[key]);
				}
			}
		}, this);
	},

	produce: function(object){
		var rawProduct = this.createRawProduct(object);
		this.refine(rawProduct);
		return rawProduct;
	}
});

module.exports = ProtoFactory;
