// router/middleware/file/index.js
var proto = require('proto');
var debug = require('debug');
var File = require('./File');
//File.prototype.aliases.push(require('./GzipAlias'));

module.exports = {
	name: 'file',
	requires: 'requestNegotiation',

	extend: {
		filePath: null,
		modifiedSinceHeaderName: 'if-modified-since',

		createFilePromise: function(path){
			var file = new File(path);

			file.fileSystem = this.fileSystem;

			var modifiedSince = this.request.getHeader(this.modifiedSinceHeaderName);
			if( typeof modifiedSince === 'string' ){
				try{
					file.modifiedSince = new Date(modifiedSince);
				}
				catch(e){
					console.warn('bad if-modified-since header value');
				}
			}
			file.acceptGzip = this.request.accept('encoding', 'gzip');

			return file.createPromise();
		}
	},

	use: function(router, path){
		router.RequestHandler.prototype.filePath = path;
	},

	// this here is router not FileHandler
	handle: function(handler){
		var request = handler.request;

		if( request.is('head', 'get') ){
			return handler.createFilePromise(handler.filePath + request.url.pathname);
		}
	}
};
