// nodetest/TestSerie/index.js
/*

*/

var TestPrototype = require('../TestPrototype');

var TestSerie = TestPrototype.create({
	Test: require('../Test'),
	type: 'serie',
	tests: null,

	init: function(tests){
		this.tests = tests;
	},

	augmentError: function(e){
		// nothing to do, the error come from a subtest
	},

	forEach: function(fn, bind){
		return this.tests.forEach(fn, bind);
	},

	createTest: function(){
		return this.Test.new.apply(this.Test, arguments);
	},

	add: function(){
		var test = this.createTest.apply(this, arguments);
		this.tests.push(test);
		return this;
	},

	setup: function(){
		this.index = 0;
		this.current = null;
	},

	teardown: function(){
		// something to do?
	},

	clean: function(){
		if( this.current ){
			this.current.stop();
		}
		TestPrototype.clean.call(this);
	},

	next: function(){
		if( this.index >= this.tests.length ){
			return null;
		}
		else{
			this.current = this.tests[this.index];
			this.index++;
			return this.current;
		}
	},

	nextTest: function(e){
		if( e ){
			this.fail(e);
		}
		else if( this.next() ){
			this.current.start(this.nextTest, this);
		}
		// all test passed with success or no test to run
		else{
			this.pass();
		}
	},

	run: function(){
		this.nextTest();
	}
});

module.exports = TestSerie;
