// router/middleware/page/index.js
/*

le comportement de renvoyer app.page au lieu de index.page ne doit pas se passer ici

MORE: 

getArguments: faudrait vérifier à la manière que global.applyScript la longeur des arguments passées / au handler

*/

var Handler = require('../handler');
var PageHandler = Handler.extend({
	Path: require('path'),
	Resolver: require('./Resolver'),
	Access: require('./Access'),
	Template: require('./Template'),

	path: null,
	module: null,
	handler: null,
	args: null,
	access: null,

	createTemplate: function(path){
		if( arguments.length === 0  ){
			// for a page in folder/page.page, search for folder/html/page.html
			path = this.Path.join(this.Path.dirname(this.path), 'html', this.Path.basename(this.path, '.page'), '.html');
		}

		var template = this.Template.new(path);

		template.router = this;
		template.send = function(params){
			var html = this.render(params), response = this.router.response;

			// on impose le content-type
			response.setHeader('content-type', 'text/html');
			response.writeBody(html);
			response.end();
		};
	},

	getModule: function(){
		var module;

		try{
			module = require(this.path);
		}
		catch(e){
			module = e;
		}

		return module;
	},

	getArguments: function(){
		var request = this.request, args, params, handler = this.handler;

		if( Array.isArray(request.bodyParams) ){
			args = request.bodyParams;
		}
		else{
			// j'ai passé du json comme paramètre, cela prévaut sur queryString
			if( params.json ){
				params = params.json;
			}
			else{
				params = request.params;
			}

			if( params == null ){
				args = [];
			}
			else if( Array.isArray(params) ){
				args = params;
			}
			else if( typeof params == 'object' ){
				args = [];
				if( typeof handler == 'function' ){
					var names = Function.argumentNames(handler), i = 0, j = names.length - 1, name;

					for(;i<j;i++){
						name = names[i];
						args.push(params[name]);
					}
				}
			}
			else{
				args = [];
			}
		}

		return args;
	},

	getHandler: function(module){
		var handler;

		// accept all method
		if( typeof module == 'function' ){
			handler = module;
		}
		else if( typeof module == 'object' ){	
			// accept all method
			if( '*' in module ){
				handler = module['*'];
			}
			else if( this.request.method in module ){
				handler = module[this.request.method];
			}

			if( typeof handler == 'object' ){
				// faut vérifier si l'url correspond au clé de l'objet, si oui on apelle les fonction en valeur
				handler = null;
			}

			if( typeof handler != 'function' ){
				handler = null;
			}
		}

		return handler;
	},

	send: function(path){
		// ceci ne doit pas se passer ici
		if( !this.request.isAjax() ){
			path = global.CLIENT_PATH + '/app.page';
		}

		this.path = path;
		this.module = this.getModule();

		// page doesn't exists
		if( this.module instanceof Error ){	
			if( this.module.code == 'MODULE_NOT_FOUND' ){
				return this.respond(404);
			}
			return this.error(this.module);
		}

		this.handler = this.getHandler(this.module);

		if( this.handler ){
			this.args = this.getArguments();
			this.access = this.Access.new(this.name);

			// this request has no right to call this page
			if( !this.access.has(this.args) ){
				return this.error(this.access.error);
			}

			try{
				this.handler.router = this.router;
				this.handler.request = this.request;
				this.handler.response = this.response;
				this.handler.page = this;

				this.handler.apply(this.handler, this.args);
			}
			catch(e){
				this.error(e);
			}
		}
		// not implemented
		else{
			this.respond(501);
		}
	},

	handle: function(router){
		this.Resolver.resolve(global.CLIENT_PATH + router.request.url.pathname, function(error, filename){
			if( filename ){
				PageHandler.new(router).send(filename);
			}
			else{
				router.next();
			}
		});
	}
});

PageHandler.response = {
	sendPage: function(path){
		return PageHandler.new(this.router).send(path);
	}
};

module.exports = PageHandler;
