// router/middleware/page/pageURLS.js
var fs = require('fs');
function getFileType(path){
	var stat;

	try{
		stat = fs.statSync(path);

		if( stat.isFile() ){
			return 'file';
		}
		if( stat.isDirectory() ){
			return 'directory';
		}
	}
	catch(e){
		return null;
	}	
}

function isFile(path){
	return getFileType(path) == 'file';
}

var PageURLS = {
	cache: {},

	getPath: function(pathname){
		if( !pathname.endsWith('.page') ) pathname+= '.page';
		return global.CLIENT_PATH + '/' + pathname;
	},

	resolve: function(pathname){
		// remove first '/'
		if( pathname[0] == '/' ) pathname = pathname.slice(1);
		if( pathname === '' ) pathname = 'index';

		var parts = pathname.split('/'), i = 0, j = parts.length, partName = '', pagePath;

		for(;i<j;i++){
			if( i !== 0 ) partName+= '/';
			partName+= parts[i];
			pagePath = this.getPath(partName);

			if( isFile(pagePath) ){
				return partName;
			}
		}

		return null;
	},

	toPage: function(pathname){
		if( pathname in this.cache ){
			return this.cache[pathname];
		}
		
		return this.cache[pathname] = this.resolve(pathname);
	},

	isPage: function(url){
		return Boolean(this.toPage(url));
	}
};

module.exports = PageURLS;
