// Table/Cell.js
var proto = require('proto');
var Cell = proto.extend({
	x: null,
	y: null,
	data: null, // row data of the cell
	value: null, // formatted value of the cell
	getter: null, // cell.data will be cell.getter()
	setter: null, // cell.value will be cell.setter(data)
	dependencies: [], // the cell.value depends of this array of cells

	meta: null, // meta concerning this cell
	fixed: false, // the cell is fixed (it cannot be sorted, moved, computed)

	constructor: function(x, y, data){
		this.x = x;
		this.y = y;
		this.data = data;
	},

	addDependency: function(cell){
		if( !Cell.isPrototypeOf(cell) ){
			throw new TypeError('A cell can depends of an other cell');
		}

		if( this.hasOwnProperty('dependencies') ){
			this.dependencies.push(cell);
		}
		else{
			this.dependencies = [cell];
		}
	},

	compute: function(){
		if( !this.fixed && this.getter ){
			this.data = this.getter.apply(this, this.dependencies.map(function(cell){
				return cell.data;
			}));
		}
	},

	format: function(){
		var data = String(this.data);

		if( !this.fixed && this.setter ){
			data = this.setter(data);
		}

		data = data.replace(/\n/g, '\\n');

		this.value = data;
		return data;
	}
});

module.exports = Cell;
