// LogStream/LogEntry/index.js
var proto = require('proto');
var LogEntry = proto.create({
	stackTrace: require('stackTrace'),
	StringTemplate: require('StringTemplate'),
	ansi: require('ansi'),
	inspect: require('util').inspect,
	fileSystem: require('fs'),

	autoReturn: true,
	depth: 2,
	showHidden: false,
	colorize: false,

	level: null,
	message: null,
	data: null,

	styles: {},

	init: function(level, message, data){
		this.level = level;
		this.message = message;
		this.data = data;
	},

	// TODO: rewrite this function
	createDebugMessage: function(message){
		/*
		var call, wrap, lineFile, lineNumber, lineContent, levelPrefix;

		call = this.stackTrace.get()[1];
		lineFile = call.getFileName();
		lineNumber = call.getFileNumber();
		lineContent = this.stackTrace.readFileLine(lineFile, lineNumber);
		lineContent = lineContent.replace(/\n|\r/, '');
		message = message.replace(/\n|\r/g, function(match){ return match + '{debug}: '; });

		wrap = '\
			{debug}: \n\
			{debug}: {lineContent} ({lineFile: {lineNumber}})\n\
			{debug}: {message}\n\
			{debug}: \
		';

		return this._render(wrap, {
			debug: 'debug',
			message: message,
			lineContent: lineContent,
			lineFile: lineFile,
			lineNumber: lineNumber
		});
		*/
		return message;
	},

	createLevelMessage: function(level, message){
		if( level === 'debug' ){
			return this.createDebugMessage(message);
		}

		if( this.colorize ){
			level = this.stylize(level, level);
		}

		return level + ': ' + message;
	},

	stylize: function(str, styleName){
		if( styleName in this.styles ){
			str = this.ansi.stylizeAll(str, this.styles[styleName]);
		}
		return str;
	},

	colorizeValue: function(value, i){
		return this.stylize(value, this.stringTemplate.expressions[i]);
	},

	toString: function(){
		this.colorize = false;

		var message = this.message, level = this.level, data = this.data, string;

		if( typeof message != 'string' ){
			message = this.inspect(message, {
				showHidden: this.showHidden,
				depth: this.depth,
				colors: this.colorize,
				customInspect: true
			});
		}

		string = this.createLevelMessage(level, message);
		if( data !== null ){
			this.stringTemplate = this.StringTemplate.new(message);
			if( this.colorize ){
				this.stringTemplate.exec(data, this.colorizeValue, this);
			}
			else{
				this.stringTemplate.exec(data);
			}

			string = this.stringTemplate.toString();
		}

		if( this.autoReturn ) string+= '\n';

		return string;
	},

	toStylizedString: function(){
		this.colorize = true;
		return this.toString();
	},

	format: function(value){
		if( typeof value == 'string' ) return value;
		try{
			return JSON.stringify(value);
		}
		catch(e){
			return '[Circular]';
		}
	},

	toJSON: function(){
		var message = this.message, level = this.level, data = this.data, jsonDataObject;

		/* create a short version of the data

		to resume calling
		logger.log('info', 'name: {name}', {name: 'damien', age:10});
		must ignore the age property to shorten the data passed to the streams
		logger.log('info', 'name: {name}', {name: 'damien', age:10});
		*/

		if( data == null ){
			jsonDataObject = null;
		}
		else{
			jsonDataObject = {};
			this.stringTemplate = this.StringTemplate.new(message);
			this.stringTemplate.exec(data, function(value, i){
				jsonDataObject[this.stringTemplate.expressions[i]] = this.format(value);
			}, this);
		}

		return {
			message: message,
			level: level,
			data: jsonDataObject
		};
	}
});

module.exports = LogEntry;
