// StringTemplate/TemplateGenerator/index.js
var proto = require('proto');
var TemplateGenerator = proto.create({
	Template: require('StringTemplate'),
	template: null,
	array: null,
	separator: null,

	init: function(template, separator){
		if( typeof template == 'string' ) template = this.Template.new(template);
		this.template = template;
		this.separator = separator;
	},

	generate: function(value, index, array){
		var template = this.template, instance;

		instance = template.new(template);		
		instance.scope = value;
		if( this.map ) instance = this.map.call(this.bind, instance, value, index, this);

		return instance;
	},

	generateAll: function(){
		var array = this.array, i = 0, j = array.length, result, templates = new Array(j);
		
		for(;i<j;i++){
			templates.push(this.generate(array[i], i, array));
		}

		return templates;
	},

	exec: function(array){
		this.array = array;
	},

	reduce: function(){
		// todo
	},

	toString: function(){
		return this.reduce('toString');
	},

	toStylizedString: function(){
		return this.reduce('toStylizedString');
	}
});

module.exports = TemplateGenerator;
