// router/middleware/bodyWriter/header/content-encoding/content-encoding.js
var AcceptableResponseHeader = require('./acceptableResponseHeader');
var ContentEncoding = AcceptableResponseHeader.extend({
	name: 'content-encoding',
	encoders: require('./encoders'),
	errorStatusCode: 500,
	errorMessage: 'Unsupported value for response header {headerName}:{headerValue}. Supported values: {supporteds}',

	isValid: function(){
		return this.value in this.encoders;
	},

	getSupporteds: function(){
		return Object.keys(this.encoders);
	},

	createError: function(){
		return AcceptableResponseHeader.createError.call(this, this.errorStatusCode, this.errorMessage, {
			supporteds: this.getSupporteds()
		});
	},

	getAccepteds: function(){
		return this.request.negotiator.encodings();
	},

	effect: function(){
		// tention si this.encoding est spécifié par le header, on suppose que body est déjà au bon format
		if( this.value && this.value != 'identity' ){
			this.response.removeHeader('content-length'); // body length is modified
			this.encoders[this.value].call(this);
		}
	},
});

module.exports = ContentEncoding;
