// unitTest/TestRunner/TestInitializer/ModuleTest/index.js
/*

sur un restart module pas besoin de recréer les series même si le module n'exporte plus la même chose
mais pour le moment on recrée tout m'en fous

*/

var TestSerie = require('../TestSerie');

var ModuleTest = TestSerie.create({
	TestSerie: TestSerie,
	type: 'moduleTest',
	testPaths: null,

	init: function(path, testPaths){
		this.path = path;
		this.setTests(testPaths);
	},

	createTest: function(path){
		return this.TestSerie.new(path, this);
	},

	getDependencyLevel: function(){
		
		function dependencyLevel(module){
			var level = 0, children = module.children, i = 0, j = children.length, child, childLevel;

			if( j ){
				for(;i<j;i++){
					child = children[i];
					childLevel = dependencyLevel(child);
					level = Math.max(level, childLevel);
				}

				level++;
			}

			return level;
		}

		return dependencyLevel(this.module);		
	},

	load: function(path){
		require(path);
		return require.cache[this.path];
	},

	clearCache: function(){
		for(var key in require.cache){
			delete require.cache[key];
		}
	},

	unload: function(path){
		this.clearCache();
	},

	reload: function(){
		this.unload(this.path);
		this.module = this.load(this.path);
		this.level = this.getDependencyLevel();
		this.tests.forEach(function(testSerie){
			testSerie.setModule(this.module);
		}, this);
	},

	// don't watch testFiles
	watchFilter: function(path){
		return this.testPaths.indexOf(path) === -1;
	}
});

module.exports = ModuleTest;
