// unitTest/AppTest/TestLoader/index.js
var TestLoader = {
	path: require('path'),
	fileSystem: require('fs'),
	testFolderName: 'test',
	testFileName: 'test.js',

	filterFolder: function(modulePath){
		var parts = modulePath.split(this.path.sep);
		return parts[parts.length - 1] !== this.testFolderName;
	},

	filterFile: function(testFile){
		return testFile.slice(-3) !== '.js';
	},

	loadFromFile: function(path){
		var testFile, dirname = this.path.dirname(path);

		testFile = dirname + this.path.sep + this.testFileName;
		if( this.fileSystem.existsSync(testFile) ){
			return testFile;
		}

		testFile = dirname + this.path.sep + this.path.basename(path, '.js') + '-test.js';
		if( this.fileSystem.existsSync(testFile) ){
			return testFile;
		}

		return false;
	},

	loadFromFolder: function(path){
		var testFolder = this.path.dirname(path) + this.path.sep + this.testFolderName;
		
		try{
			// any files in it is a test
			return this.fileSystem.readdirSync(filterFile).filter(this.filterFile, this).map(function(name){
				return testFolder + this.path.sep + name;
			}, this);
		}
		catch(e){
			return false;
		}
	},

	load: function(modulePath){
		if( this.filterFolder(modulePath) ){
			return this.loadFromFile(modulePath) || this.loadFromFolder(modulePath);
		}
		else{
			return false;
		}
	}
};

module.exports = TestLoader;
